// solana/bin/index.mjs
#!/usr/bin/env node
import require$$0$1 from "node:events";

import require$$1 from "node:child_process";

import require$$2 from "node:path";

import require$$3 from "node:fs";

import process$2 from "node:process";

import os from "node:os";

import fs from "fs";

import path from "path";

import * as childProcess from "child_process";

import tty from "node:tty";

function getDefaultExportFromCjs(x) {
    return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, "default") ? x.default : x;
}

var hasRequiredPicocolors, picocolors$1 = {
    exports: {}
};

function requirePicocolors() {
    if (hasRequiredPicocolors) return picocolors$1.exports;
    hasRequiredPicocolors = 1;
    let p = process || {}, argv = p.argv || [], env = p.env || {}, isColorSupported = !(env.NO_COLOR || argv.includes("--no-color")) && (!!env.FORCE_COLOR || argv.includes("--color") || "win32" === p.platform || (p.stdout || {}).isTTY && "dumb" !== env.TERM || !!env.CI), formatter = (open, close, replace = open) => input => {
        let string = "" + input, index = string.indexOf(close, open.length);
        return ~index ? open + replaceClose(string, close, replace, index) + close : open + string + close;
    }, replaceClose = (string, close, replace, index) => {
        let result = "", cursor = 0;
        do {
            result += string.substring(cursor, index) + replace, cursor = index + close.length, 
            index = string.indexOf(close, cursor);
        } while (~index);
        return result + string.substring(cursor);
    }, createColors = (enabled = isColorSupported) => {
        let f = enabled ? formatter : () => String;
        return {
            isColorSupported: enabled,
            reset: f("[0m", "[0m"),
            bold: f("[1m", "[22m", "[22m[1m"),
            dim: f("[2m", "[22m", "[22m[2m"),
            italic: f("[3m", "[23m"),
            underline: f("[4m", "[24m"),
            inverse: f("[7m", "[27m"),
            hidden: f("[8m", "[28m"),
            strikethrough: f("[9m", "[29m"),
            black: f("[30m", "[39m"),
            red: f("[31m", "[39m"),
            green: f("[32m", "[39m"),
            yellow: f("[33m", "[39m"),
            blue: f("[34m", "[39m"),
            magenta: f("[35m", "[39m"),
            cyan: f("[36m", "[39m"),
            white: f("[37m", "[39m"),
            gray: f("[90m", "[39m"),
            bgBlack: f("[40m", "[49m"),
            bgRed: f("[41m", "[49m"),
            bgGreen: f("[42m", "[49m"),
            bgYellow: f("[43m", "[49m"),
            bgBlue: f("[44m", "[49m"),
            bgMagenta: f("[45m", "[49m"),
            bgCyan: f("[46m", "[49m"),
            bgWhite: f("[47m", "[49m"),
            blackBright: f("[90m", "[39m"),
            redBright: f("[91m", "[39m"),
            greenBright: f("[92m", "[39m"),
            yellowBright: f("[93m", "[39m"),
            blueBright: f("[94m", "[39m"),
            magentaBright: f("[95m", "[39m"),
            cyanBright: f("[96m", "[39m"),
            whiteBright: f("[97m", "[39m"),
            bgBlackBright: f("[100m", "[49m"),
            bgRedBright: f("[101m", "[49m"),
            bgGreenBright: f("[102m", "[49m"),
            bgYellowBright: f("[103m", "[49m"),
            bgBlueBright: f("[104m", "[49m"),
            bgMagentaBright: f("[105m", "[49m"),
            bgCyanBright: f("[106m", "[49m"),
            bgWhiteBright: f("[107m", "[49m")
        };
    };
    return picocolors$1.exports = createColors(), picocolors$1.exports.createColors = createColors, 
    picocolors$1.exports;
}

var picocolors = getDefaultExportFromCjs(requirePicocolors());

function warnMessage(msg, highlight = !1) {
    console.log(picocolors.yellow(msg));
}

function errorMessage(err, title = "An error occurred") {
    let message = "Unknown error";
    "string" == typeof err ? message = err : err instanceof Error && (message = err.message), 
    function(msg, title = null) {
        title ? (console.log(picocolors.bgRed(` ${title} `)), console.log(msg, "\n")) : console.log(picocolors.bgRed(` ${msg} `), "\n"), 
        process.exit(1);
    }(message, title);
}

var hasRequiredError, hasRequiredArgument, extraTypings = {
    exports: {}
}, commander = {}, argument = {}, error = {};

function requireError() {
    if (hasRequiredError) return error;
    hasRequiredError = 1;
    class CommanderError extends Error {
        constructor(exitCode, code, message) {
            super(message), Error.captureStackTrace(this, this.constructor), this.name = this.constructor.name, 
            this.code = code, this.exitCode = exitCode, this.nestedError = void 0;
        }
    }
    return error.CommanderError = CommanderError, error.InvalidArgumentError = class extends CommanderError {
        constructor(message) {
            super(1, "commander.invalidArgument", message), Error.captureStackTrace(this, this.constructor), 
            this.name = this.constructor.name;
        }
    }, error;
}

function requireArgument() {
    if (hasRequiredArgument) return argument;
    hasRequiredArgument = 1;
    const {InvalidArgumentError: InvalidArgumentError} = requireError();
    return argument.Argument = class {
        constructor(name, description) {
            switch (this.description = description || "", this.variadic = !1, this.parseArg = void 0, 
            this.defaultValue = void 0, this.defaultValueDescription = void 0, this.argChoices = void 0, 
            name[0]) {
              case "<":
                this.required = !0, this._name = name.slice(1, -1);
                break;

              case "[":
                this.required = !1, this._name = name.slice(1, -1);
                break;

              default:
                this.required = !0, this._name = name;
            }
            this._name.length > 3 && "..." === this._name.slice(-3) && (this.variadic = !0, 
            this._name = this._name.slice(0, -3));
        }
        name() {
            return this._name;
        }
        _concatValue(value, previous) {
            return previous !== this.defaultValue && Array.isArray(previous) ? previous.concat(value) : [ value ];
        }
        default(value, description) {
            return this.defaultValue = value, this.defaultValueDescription = description, this;
        }
        argParser(fn) {
            return this.parseArg = fn, this;
        }
        choices(values) {
            return this.argChoices = values.slice(), this.parseArg = (arg, previous) => {
                if (!this.argChoices.includes(arg)) throw new InvalidArgumentError(`Allowed choices are ${this.argChoices.join(", ")}.`);
                return this.variadic ? this._concatValue(arg, previous) : arg;
            }, this;
        }
        argRequired() {
            return this.required = !0, this;
        }
        argOptional() {
            return this.required = !1, this;
        }
    }, argument.humanReadableArgName = function(arg) {
        const nameOutput = arg.name() + (!0 === arg.variadic ? "..." : "");
        return arg.required ? "<" + nameOutput + ">" : "[" + nameOutput + "]";
    }, argument;
}

var hasRequiredHelp, command = {}, help = {};

function requireHelp() {
    if (hasRequiredHelp) return help;
    hasRequiredHelp = 1;
    const {humanReadableArgName: humanReadableArgName} = requireArgument();
    return help.Help = class {
        constructor() {
            this.helpWidth = void 0, this.sortSubcommands = !1, this.sortOptions = !1, this.showGlobalOptions = !1;
        }
        visibleCommands(cmd) {
            const visibleCommands = cmd.commands.filter((cmd => !cmd._hidden)), helpCommand = cmd._getHelpCommand();
            return helpCommand && !helpCommand._hidden && visibleCommands.push(helpCommand), 
            this.sortSubcommands && visibleCommands.sort(((a, b) => a.name().localeCompare(b.name()))), 
            visibleCommands;
        }
        compareOptions(a, b) {
            const getSortKey = option => option.short ? option.short.replace(/^-/, "") : option.long.replace(/^--/, "");
            return getSortKey(a).localeCompare(getSortKey(b));
        }
        visibleOptions(cmd) {
            const visibleOptions = cmd.options.filter((option => !option.hidden)), helpOption = cmd._getHelpOption();
            if (helpOption && !helpOption.hidden) {
                const removeShort = helpOption.short && cmd._findOption(helpOption.short), removeLong = helpOption.long && cmd._findOption(helpOption.long);
                removeShort || removeLong ? helpOption.long && !removeLong ? visibleOptions.push(cmd.createOption(helpOption.long, helpOption.description)) : helpOption.short && !removeShort && visibleOptions.push(cmd.createOption(helpOption.short, helpOption.description)) : visibleOptions.push(helpOption);
            }
            return this.sortOptions && visibleOptions.sort(this.compareOptions), visibleOptions;
        }
        visibleGlobalOptions(cmd) {
            if (!this.showGlobalOptions) return [];
            const globalOptions = [];
            for (let ancestorCmd = cmd.parent; ancestorCmd; ancestorCmd = ancestorCmd.parent) {
                const visibleOptions = ancestorCmd.options.filter((option => !option.hidden));
                globalOptions.push(...visibleOptions);
            }
            return this.sortOptions && globalOptions.sort(this.compareOptions), globalOptions;
        }
        visibleArguments(cmd) {
            return cmd._argsDescription && cmd.registeredArguments.forEach((argument => {
                argument.description = argument.description || cmd._argsDescription[argument.name()] || "";
            })), cmd.registeredArguments.find((argument => argument.description)) ? cmd.registeredArguments : [];
        }
        subcommandTerm(cmd) {
            const args = cmd.registeredArguments.map((arg => humanReadableArgName(arg))).join(" ");
            return cmd._name + (cmd._aliases[0] ? "|" + cmd._aliases[0] : "") + (cmd.options.length ? " [options]" : "") + (args ? " " + args : "");
        }
        optionTerm(option) {
            return option.flags;
        }
        argumentTerm(argument) {
            return argument.name();
        }
        longestSubcommandTermLength(cmd, helper) {
            return helper.visibleCommands(cmd).reduce(((max, command) => Math.max(max, helper.subcommandTerm(command).length)), 0);
        }
        longestOptionTermLength(cmd, helper) {
            return helper.visibleOptions(cmd).reduce(((max, option) => Math.max(max, helper.optionTerm(option).length)), 0);
        }
        longestGlobalOptionTermLength(cmd, helper) {
            return helper.visibleGlobalOptions(cmd).reduce(((max, option) => Math.max(max, helper.optionTerm(option).length)), 0);
        }
        longestArgumentTermLength(cmd, helper) {
            return helper.visibleArguments(cmd).reduce(((max, argument) => Math.max(max, helper.argumentTerm(argument).length)), 0);
        }
        commandUsage(cmd) {
            let cmdName = cmd._name;
            cmd._aliases[0] && (cmdName = cmdName + "|" + cmd._aliases[0]);
            let ancestorCmdNames = "";
            for (let ancestorCmd = cmd.parent; ancestorCmd; ancestorCmd = ancestorCmd.parent) ancestorCmdNames = ancestorCmd.name() + " " + ancestorCmdNames;
            return ancestorCmdNames + cmdName + " " + cmd.usage();
        }
        commandDescription(cmd) {
            return cmd.description();
        }
        subcommandDescription(cmd) {
            return cmd.summary() || cmd.description();
        }
        optionDescription(option) {
            const extraInfo = [];
            if (option.argChoices && extraInfo.push(`choices: ${option.argChoices.map((choice => JSON.stringify(choice))).join(", ")}`), 
            void 0 !== option.defaultValue) {
                (option.required || option.optional || option.isBoolean() && "boolean" == typeof option.defaultValue) && extraInfo.push(`default: ${option.defaultValueDescription || JSON.stringify(option.defaultValue)}`);
            }
            return void 0 !== option.presetArg && option.optional && extraInfo.push(`preset: ${JSON.stringify(option.presetArg)}`), 
            void 0 !== option.envVar && extraInfo.push(`env: ${option.envVar}`), extraInfo.length > 0 ? `${option.description} (${extraInfo.join(", ")})` : option.description;
        }
        argumentDescription(argument) {
            const extraInfo = [];
            if (argument.argChoices && extraInfo.push(`choices: ${argument.argChoices.map((choice => JSON.stringify(choice))).join(", ")}`), 
            void 0 !== argument.defaultValue && extraInfo.push(`default: ${argument.defaultValueDescription || JSON.stringify(argument.defaultValue)}`), 
            extraInfo.length > 0) {
                const extraDescripton = `(${extraInfo.join(", ")})`;
                return argument.description ? `${argument.description} ${extraDescripton}` : extraDescripton;
            }
            return argument.description;
        }
        formatHelp(cmd, helper) {
            const termWidth = helper.padWidth(cmd, helper), helpWidth = helper.helpWidth || 80;
            function formatItem(term, description) {
                if (description) {
                    const fullText = `${term.padEnd(termWidth + 2)}${description}`;
                    return helper.wrap(fullText, helpWidth - 2, termWidth + 2);
                }
                return term;
            }
            function formatList(textArray) {
                return textArray.join("\n").replace(/^/gm, " ".repeat(2));
            }
            let output = [ `Usage: ${helper.commandUsage(cmd)}`, "" ];
            const commandDescription = helper.commandDescription(cmd);
            commandDescription.length > 0 && (output = output.concat([ helper.wrap(commandDescription, helpWidth, 0), "" ]));
            const argumentList = helper.visibleArguments(cmd).map((argument => formatItem(helper.argumentTerm(argument), helper.argumentDescription(argument))));
            argumentList.length > 0 && (output = output.concat([ "Arguments:", formatList(argumentList), "" ]));
            const optionList = helper.visibleOptions(cmd).map((option => formatItem(helper.optionTerm(option), helper.optionDescription(option))));
            if (optionList.length > 0 && (output = output.concat([ "Options:", formatList(optionList), "" ])), 
            this.showGlobalOptions) {
                const globalOptionList = helper.visibleGlobalOptions(cmd).map((option => formatItem(helper.optionTerm(option), helper.optionDescription(option))));
                globalOptionList.length > 0 && (output = output.concat([ "Global Options:", formatList(globalOptionList), "" ]));
            }
            const commandList = helper.visibleCommands(cmd).map((cmd => formatItem(helper.subcommandTerm(cmd), helper.subcommandDescription(cmd))));
            return commandList.length > 0 && (output = output.concat([ "Commands:", formatList(commandList), "" ])), 
            output.join("\n");
        }
        padWidth(cmd, helper) {
            return Math.max(helper.longestOptionTermLength(cmd, helper), helper.longestGlobalOptionTermLength(cmd, helper), helper.longestSubcommandTermLength(cmd, helper), helper.longestArgumentTermLength(cmd, helper));
        }
        wrap(str, width, indent, minColumnWidth = 40) {
            const manualIndent = new RegExp("[\\n][ \\f\\t\\vТ ст-ттЏту\ufeff]+");
            if (str.match(manualIndent)) return str;
            const columnWidth = width - indent;
            if (columnWidth < minColumnWidth) return str;
            const leadingStr = str.slice(0, indent), columnText = str.slice(indent).replace("\r\n", "\n"), indentString = " ".repeat(indent), regex = new RegExp(`\n|.{1,${columnWidth - 1}}([\\sт]|$)|[^\\sт]+?([\\sт]|$)`, "g");
            return leadingStr + (columnText.match(regex) || []).map(((line, i) => "\n" === line ? "" : (i > 0 ? indentString : "") + line.trimEnd())).join("\n");
        }
    }, help;
}

var hasRequiredOption, option = {};

function requireOption() {
    if (hasRequiredOption) return option;
    hasRequiredOption = 1;
    const {InvalidArgumentError: InvalidArgumentError} = requireError();
    return option.Option = class {
        constructor(flags, description) {
            this.flags = flags, this.description = description || "", this.required = flags.includes("<"), 
            this.optional = flags.includes("["), this.variadic = /\w\.\.\.[>\]]$/.test(flags), 
            this.mandatory = !1;
            const optionFlags = function(flags) {
                let shortFlag, longFlag;
                const flagParts = flags.split(/[ |,]+/);
                flagParts.length > 1 && !/^[[<]/.test(flagParts[1]) && (shortFlag = flagParts.shift());
                longFlag = flagParts.shift(), !shortFlag && /^-[^-]$/.test(longFlag) && (shortFlag = longFlag, 
                longFlag = void 0);
                return {
                    shortFlag: shortFlag,
                    longFlag: longFlag
                };
            }(flags);
            this.short = optionFlags.shortFlag, this.long = optionFlags.longFlag, this.negate = !1, 
            this.long && (this.negate = this.long.startsWith("--no-")), this.defaultValue = void 0, 
            this.defaultValueDescription = void 0, this.presetArg = void 0, this.envVar = void 0, 
            this.parseArg = void 0, this.hidden = !1, this.argChoices = void 0, this.conflictsWith = [], 
            this.implied = void 0;
        }
        default(value, description) {
            return this.defaultValue = value, this.defaultValueDescription = description, this;
        }
        preset(arg) {
            return this.presetArg = arg, this;
        }
        conflicts(names) {
            return this.conflictsWith = this.conflictsWith.concat(names), this;
        }
        implies(impliedOptionValues) {
            let newImplied = impliedOptionValues;
            return "string" == typeof impliedOptionValues && (newImplied = {
                [impliedOptionValues]: !0
            }), this.implied = Object.assign(this.implied || {}, newImplied), this;
        }
        env(name) {
            return this.envVar = name, this;
        }
        argParser(fn) {
            return this.parseArg = fn, this;
        }
        makeOptionMandatory(mandatory = !0) {
            return this.mandatory = !!mandatory, this;
        }
        hideHelp(hide = !0) {
            return this.hidden = !!hide, this;
        }
        _concatValue(value, previous) {
            return previous !== this.defaultValue && Array.isArray(previous) ? previous.concat(value) : [ value ];
        }
        choices(values) {
            return this.argChoices = values.slice(), this.parseArg = (arg, previous) => {
                if (!this.argChoices.includes(arg)) throw new InvalidArgumentError(`Allowed choices are ${this.argChoices.join(", ")}.`);
                return this.variadic ? this._concatValue(arg, previous) : arg;
            }, this;
        }
        name() {
            return this.long ? this.long.replace(/^--/, "") : this.short.replace(/^-/, "");
        }
        attributeName() {
            return this.name().replace(/^no-/, "").split("-").reduce(((str, word) => str + word[0].toUpperCase() + word.slice(1)));
        }
        is(arg) {
            return this.short === arg || this.long === arg;
        }
        isBoolean() {
            return !this.required && !this.optional && !this.negate;
        }
    }, option.DualOptions = class {
        constructor(options) {
            this.positiveOptions = new Map, this.negativeOptions = new Map, this.dualOptions = new Set, 
            options.forEach((option => {
                option.negate ? this.negativeOptions.set(option.attributeName(), option) : this.positiveOptions.set(option.attributeName(), option);
            })), this.negativeOptions.forEach(((value, key) => {
                this.positiveOptions.has(key) && this.dualOptions.add(key);
            }));
        }
        valueFromOption(value, option) {
            const optionKey = option.attributeName();
            if (!this.dualOptions.has(optionKey)) return !0;
            const preset = this.negativeOptions.get(optionKey).presetArg, negativeValue = void 0 !== preset && preset;
            return option.negate === (negativeValue === value);
        }
    }, option;
}

var hasRequiredSuggestSimilar, hasRequiredCommand, hasRequiredCommander, hasRequiredExtraTypings, suggestSimilar = {};

function requireSuggestSimilar() {
    if (hasRequiredSuggestSimilar) return suggestSimilar;
    hasRequiredSuggestSimilar = 1;
    return suggestSimilar.suggestSimilar = function(word, candidates) {
        if (!candidates || 0 === candidates.length) return "";
        candidates = Array.from(new Set(candidates));
        const searchingOptions = word.startsWith("--");
        searchingOptions && (word = word.slice(2), candidates = candidates.map((candidate => candidate.slice(2))));
        let similar = [], bestDistance = 3;
        return candidates.forEach((candidate => {
            if (candidate.length <= 1) return;
            const distance = function(a, b) {
                if (Math.abs(a.length - b.length) > 3) return Math.max(a.length, b.length);
                const d = [];
                for (let i = 0; i <= a.length; i++) d[i] = [ i ];
                for (let j = 0; j <= b.length; j++) d[0][j] = j;
                for (let j = 1; j <= b.length; j++) for (let i = 1; i <= a.length; i++) {
                    let cost = 1;
                    cost = a[i - 1] === b[j - 1] ? 0 : 1, d[i][j] = Math.min(d[i - 1][j] + 1, d[i][j - 1] + 1, d[i - 1][j - 1] + cost), 
                    i > 1 && j > 1 && a[i - 1] === b[j - 2] && a[i - 2] === b[j - 1] && (d[i][j] = Math.min(d[i][j], d[i - 2][j - 2] + 1));
                }
                return d[a.length][b.length];
            }(word, candidate), length = Math.max(word.length, candidate.length);
            (length - distance) / length > .4 && (distance < bestDistance ? (bestDistance = distance, 
            similar = [ candidate ]) : distance === bestDistance && similar.push(candidate));
        })), similar.sort(((a, b) => a.localeCompare(b))), searchingOptions && (similar = similar.map((candidate => `--${candidate}`))), 
        similar.length > 1 ? `\n(Did you mean one of ${similar.join(", ")}?)` : 1 === similar.length ? `\n(Did you mean ${similar[0]}?)` : "";
    }, suggestSimilar;
}

function requireCommander() {
    if (hasRequiredCommander) return commander;
    hasRequiredCommander = 1;
    const {Argument: Argument} = requireArgument(), {Command: Command} = function() {
        if (hasRequiredCommand) return command;
        hasRequiredCommand = 1;
        const EventEmitter = require$$0$1.EventEmitter, childProcess = require$$1, path = require$$2, fs = require$$3, process = process$2, {Argument: Argument, humanReadableArgName: humanReadableArgName} = requireArgument(), {CommanderError: CommanderError} = requireError(), {Help: Help} = requireHelp(), {Option: Option, DualOptions: DualOptions} = requireOption(), {suggestSimilar: suggestSimilar} = requireSuggestSimilar();
        class Command extends EventEmitter {
            constructor(name) {
                super(), this.commands = [], this.options = [], this.parent = null, this._allowUnknownOption = !1, 
                this._allowExcessArguments = !0, this.registeredArguments = [], this._args = this.registeredArguments, 
                this.args = [], this.rawArgs = [], this.processedArgs = [], this._scriptPath = null, 
                this._name = name || "", this._optionValues = {}, this._optionValueSources = {}, 
                this._storeOptionsAsProperties = !1, this._actionHandler = null, this._executableHandler = !1, 
                this._executableFile = null, this._executableDir = null, this._defaultCommandName = null, 
                this._exitCallback = null, this._aliases = [], this._combineFlagAndOptionalValue = !0, 
                this._description = "", this._summary = "", this._argsDescription = void 0, this._enablePositionalOptions = !1, 
                this._passThroughOptions = !1, this._lifeCycleHooks = {}, this._showHelpAfterError = !1, 
                this._showSuggestionAfterError = !0, this._outputConfiguration = {
                    writeOut: str => process.stdout.write(str),
                    writeErr: str => process.stderr.write(str),
                    getOutHelpWidth: () => process.stdout.isTTY ? process.stdout.columns : void 0,
                    getErrHelpWidth: () => process.stderr.isTTY ? process.stderr.columns : void 0,
                    outputError: (str, write) => write(str)
                }, this._hidden = !1, this._helpOption = void 0, this._addImplicitHelpCommand = void 0, 
                this._helpCommand = void 0, this._helpConfiguration = {};
            }
            copyInheritedSettings(sourceCommand) {
                return this._outputConfiguration = sourceCommand._outputConfiguration, this._helpOption = sourceCommand._helpOption, 
                this._helpCommand = sourceCommand._helpCommand, this._helpConfiguration = sourceCommand._helpConfiguration, 
                this._exitCallback = sourceCommand._exitCallback, this._storeOptionsAsProperties = sourceCommand._storeOptionsAsProperties, 
                this._combineFlagAndOptionalValue = sourceCommand._combineFlagAndOptionalValue, 
                this._allowExcessArguments = sourceCommand._allowExcessArguments, this._enablePositionalOptions = sourceCommand._enablePositionalOptions, 
                this._showHelpAfterError = sourceCommand._showHelpAfterError, this._showSuggestionAfterError = sourceCommand._showSuggestionAfterError, 
                this;
            }
            _getCommandAndAncestors() {
                const result = [];
                for (let command = this; command; command = command.parent) result.push(command);
                return result;
            }
            command(nameAndArgs, actionOptsOrExecDesc, execOpts) {
                let desc = actionOptsOrExecDesc, opts = execOpts;
                "object" == typeof desc && null !== desc && (opts = desc, desc = null), opts = opts || {};
                const [, name, args] = nameAndArgs.match(/([^ ]+) *(.*)/), cmd = this.createCommand(name);
                return desc && (cmd.description(desc), cmd._executableHandler = !0), opts.isDefault && (this._defaultCommandName = cmd._name), 
                cmd._hidden = !(!opts.noHelp && !opts.hidden), cmd._executableFile = opts.executableFile || null, 
                args && cmd.arguments(args), this._registerCommand(cmd), cmd.parent = this, cmd.copyInheritedSettings(this), 
                desc ? this : cmd;
            }
            createCommand(name) {
                return new Command(name);
            }
            createHelp() {
                return Object.assign(new Help, this.configureHelp());
            }
            configureHelp(configuration) {
                return void 0 === configuration ? this._helpConfiguration : (this._helpConfiguration = configuration, 
                this);
            }
            configureOutput(configuration) {
                return void 0 === configuration ? this._outputConfiguration : (Object.assign(this._outputConfiguration, configuration), 
                this);
            }
            showHelpAfterError(displayHelp = !0) {
                return "string" != typeof displayHelp && (displayHelp = !!displayHelp), this._showHelpAfterError = displayHelp, 
                this;
            }
            showSuggestionAfterError(displaySuggestion = !0) {
                return this._showSuggestionAfterError = !!displaySuggestion, this;
            }
            addCommand(cmd, opts) {
                if (!cmd._name) throw new Error("Command passed to .addCommand() must have a name\n- specify the name in Command constructor or using .name()");
                return (opts = opts || {}).isDefault && (this._defaultCommandName = cmd._name), 
                (opts.noHelp || opts.hidden) && (cmd._hidden = !0), this._registerCommand(cmd), 
                cmd.parent = this, cmd._checkForBrokenPassThrough(), this;
            }
            createArgument(name, description) {
                return new Argument(name, description);
            }
            argument(name, description, fn, defaultValue) {
                const argument = this.createArgument(name, description);
                return "function" == typeof fn ? argument.default(defaultValue).argParser(fn) : argument.default(fn), 
                this.addArgument(argument), this;
            }
            arguments(names) {
                return names.trim().split(/ +/).forEach((detail => {
                    this.argument(detail);
                })), this;
            }
            addArgument(argument) {
                const previousArgument = this.registeredArguments.slice(-1)[0];
                if (previousArgument && previousArgument.variadic) throw new Error(`only the last argument can be variadic '${previousArgument.name()}'`);
                if (argument.required && void 0 !== argument.defaultValue && void 0 === argument.parseArg) throw new Error(`a default value for a required argument is never used: '${argument.name()}'`);
                return this.registeredArguments.push(argument), this;
            }
            helpCommand(enableOrNameAndArgs, description) {
                if ("boolean" == typeof enableOrNameAndArgs) return this._addImplicitHelpCommand = enableOrNameAndArgs, 
                this;
                enableOrNameAndArgs = enableOrNameAndArgs ?? "help [command]";
                const [, helpName, helpArgs] = enableOrNameAndArgs.match(/([^ ]+) *(.*)/), helpDescription = description ?? "display help for command", helpCommand = this.createCommand(helpName);
                return helpCommand.helpOption(!1), helpArgs && helpCommand.arguments(helpArgs), 
                helpDescription && helpCommand.description(helpDescription), this._addImplicitHelpCommand = !0, 
                this._helpCommand = helpCommand, this;
            }
            addHelpCommand(helpCommand, deprecatedDescription) {
                return "object" != typeof helpCommand ? (this.helpCommand(helpCommand, deprecatedDescription), 
                this) : (this._addImplicitHelpCommand = !0, this._helpCommand = helpCommand, this);
            }
            _getHelpCommand() {
                return this._addImplicitHelpCommand ?? (this.commands.length && !this._actionHandler && !this._findCommand("help")) ? (void 0 === this._helpCommand && this.helpCommand(void 0, void 0), 
                this._helpCommand) : null;
            }
            hook(event, listener) {
                const allowedValues = [ "preSubcommand", "preAction", "postAction" ];
                if (!allowedValues.includes(event)) throw new Error(`Unexpected value for event passed to hook : '${event}'.\nExpecting one of '${allowedValues.join("', '")}'`);
                return this._lifeCycleHooks[event] ? this._lifeCycleHooks[event].push(listener) : this._lifeCycleHooks[event] = [ listener ], 
                this;
            }
            exitOverride(fn) {
                return this._exitCallback = fn || (err => {
                    if ("commander.executeSubCommandAsync" !== err.code) throw err;
                }), this;
            }
            _exit(exitCode, code, message) {
                this._exitCallback && this._exitCallback(new CommanderError(exitCode, code, message)), 
                process.exit(exitCode);
            }
            action(fn) {
                return this._actionHandler = args => {
                    const expectedArgsCount = this.registeredArguments.length, actionArgs = args.slice(0, expectedArgsCount);
                    return this._storeOptionsAsProperties ? actionArgs[expectedArgsCount] = this : actionArgs[expectedArgsCount] = this.opts(), 
                    actionArgs.push(this), fn.apply(this, actionArgs);
                }, this;
            }
            createOption(flags, description) {
                return new Option(flags, description);
            }
            _callParseArg(target, value, previous, invalidArgumentMessage) {
                try {
                    return target.parseArg(value, previous);
                } catch (err) {
                    if ("commander.invalidArgument" === err.code) {
                        const message = `${invalidArgumentMessage} ${err.message}`;
                        this.error(message, {
                            exitCode: err.exitCode,
                            code: err.code
                        });
                    }
                    throw err;
                }
            }
            _registerOption(option) {
                const matchingOption = option.short && this._findOption(option.short) || option.long && this._findOption(option.long);
                if (matchingOption) {
                    const matchingFlag = option.long && this._findOption(option.long) ? option.long : option.short;
                    throw new Error(`Cannot add option '${option.flags}'${this._name && ` to command '${this._name}'`} due to conflicting flag '${matchingFlag}'\n-  already used by option '${matchingOption.flags}'`);
                }
                this.options.push(option);
            }
            _registerCommand(command) {
                const knownBy = cmd => [ cmd.name() ].concat(cmd.aliases()), alreadyUsed = knownBy(command).find((name => this._findCommand(name)));
                if (alreadyUsed) {
                    const existingCmd = knownBy(this._findCommand(alreadyUsed)).join("|"), newCmd = knownBy(command).join("|");
                    throw new Error(`cannot add command '${newCmd}' as already have command '${existingCmd}'`);
                }
                this.commands.push(command);
            }
            addOption(option) {
                this._registerOption(option);
                const oname = option.name(), name = option.attributeName();
                if (option.negate) {
                    const positiveLongFlag = option.long.replace(/^--no-/, "--");
                    this._findOption(positiveLongFlag) || this.setOptionValueWithSource(name, void 0 === option.defaultValue || option.defaultValue, "default");
                } else void 0 !== option.defaultValue && this.setOptionValueWithSource(name, option.defaultValue, "default");
                const handleOptionValue = (val, invalidValueMessage, valueSource) => {
                    null == val && void 0 !== option.presetArg && (val = option.presetArg);
                    const oldValue = this.getOptionValue(name);
                    null !== val && option.parseArg ? val = this._callParseArg(option, val, oldValue, invalidValueMessage) : null !== val && option.variadic && (val = option._concatValue(val, oldValue)), 
                    null == val && (val = !option.negate && (!(!option.isBoolean() && !option.optional) || "")), 
                    this.setOptionValueWithSource(name, val, valueSource);
                };
                return this.on("option:" + oname, (val => {
                    const invalidValueMessage = `error: option '${option.flags}' argument '${val}' is invalid.`;
                    handleOptionValue(val, invalidValueMessage, "cli");
                })), option.envVar && this.on("optionEnv:" + oname, (val => {
                    const invalidValueMessage = `error: option '${option.flags}' value '${val}' from env '${option.envVar}' is invalid.`;
                    handleOptionValue(val, invalidValueMessage, "env");
                })), this;
            }
            _optionEx(config, flags, description, fn, defaultValue) {
                if ("object" == typeof flags && flags instanceof Option) throw new Error("To add an Option object use addOption() instead of option() or requiredOption()");
                const option = this.createOption(flags, description);
                if (option.makeOptionMandatory(!!config.mandatory), "function" == typeof fn) option.default(defaultValue).argParser(fn); else if (fn instanceof RegExp) {
                    const regex = fn;
                    fn = (val, def) => {
                        const m = regex.exec(val);
                        return m ? m[0] : def;
                    }, option.default(defaultValue).argParser(fn);
                } else option.default(fn);
                return this.addOption(option);
            }
            option(flags, description, parseArg, defaultValue) {
                return this._optionEx({}, flags, description, parseArg, defaultValue);
            }
            requiredOption(flags, description, parseArg, defaultValue) {
                return this._optionEx({
                    mandatory: !0
                }, flags, description, parseArg, defaultValue);
            }
            combineFlagAndOptionalValue(combine = !0) {
                return this._combineFlagAndOptionalValue = !!combine, this;
            }
            allowUnknownOption(allowUnknown = !0) {
                return this._allowUnknownOption = !!allowUnknown, this;
            }
            allowExcessArguments(allowExcess = !0) {
                return this._allowExcessArguments = !!allowExcess, this;
            }
            enablePositionalOptions(positional = !0) {
                return this._enablePositionalOptions = !!positional, this;
            }
            passThroughOptions(passThrough = !0) {
                return this._passThroughOptions = !!passThrough, this._checkForBrokenPassThrough(), 
                this;
            }
            _checkForBrokenPassThrough() {
                if (this.parent && this._passThroughOptions && !this.parent._enablePositionalOptions) throw new Error(`passThroughOptions cannot be used for '${this._name}' without turning on enablePositionalOptions for parent command(s)`);
            }
            storeOptionsAsProperties(storeAsProperties = !0) {
                if (this.options.length) throw new Error("call .storeOptionsAsProperties() before adding options");
                if (Object.keys(this._optionValues).length) throw new Error("call .storeOptionsAsProperties() before setting option values");
                return this._storeOptionsAsProperties = !!storeAsProperties, this;
            }
            getOptionValue(key) {
                return this._storeOptionsAsProperties ? this[key] : this._optionValues[key];
            }
            setOptionValue(key, value) {
                return this.setOptionValueWithSource(key, value, void 0);
            }
            setOptionValueWithSource(key, value, source) {
                return this._storeOptionsAsProperties ? this[key] = value : this._optionValues[key] = value, 
                this._optionValueSources[key] = source, this;
            }
            getOptionValueSource(key) {
                return this._optionValueSources[key];
            }
            getOptionValueSourceWithGlobals(key) {
                let source;
                return this._getCommandAndAncestors().forEach((cmd => {
                    void 0 !== cmd.getOptionValueSource(key) && (source = cmd.getOptionValueSource(key));
                })), source;
            }
            _prepareUserArgs(argv, parseOptions) {
                if (void 0 !== argv && !Array.isArray(argv)) throw new Error("first parameter to parse must be array or undefined");
                if (parseOptions = parseOptions || {}, void 0 === argv && void 0 === parseOptions.from) {
                    process.versions?.electron && (parseOptions.from = "electron");
                    const execArgv = process.execArgv ?? [];
                    (execArgv.includes("-e") || execArgv.includes("--eval") || execArgv.includes("-p") || execArgv.includes("--print")) && (parseOptions.from = "eval");
                }
                let userArgs;
                switch (void 0 === argv && (argv = process.argv), this.rawArgs = argv.slice(), parseOptions.from) {
                  case void 0:
                  case "node":
                    this._scriptPath = argv[1], userArgs = argv.slice(2);
                    break;

                  case "electron":
                    process.defaultApp ? (this._scriptPath = argv[1], userArgs = argv.slice(2)) : userArgs = argv.slice(1);
                    break;

                  case "user":
                    userArgs = argv.slice(0);
                    break;

                  case "eval":
                    userArgs = argv.slice(1);
                    break;

                  default:
                    throw new Error(`unexpected parse option { from: '${parseOptions.from}' }`);
                }
                return !this._name && this._scriptPath && this.nameFromFilename(this._scriptPath), 
                this._name = this._name || "program", userArgs;
            }
            parse(argv, parseOptions) {
                const userArgs = this._prepareUserArgs(argv, parseOptions);
                return this._parseCommand([], userArgs), this;
            }
            async parseAsync(argv, parseOptions) {
                const userArgs = this._prepareUserArgs(argv, parseOptions);
                return await this._parseCommand([], userArgs), this;
            }
            _executeSubCommand(subcommand, args) {
                args = args.slice();
                let launchWithNode = !1;
                const sourceExt = [ ".js", ".ts", ".tsx", ".mjs", ".cjs" ];
                function findFile(baseDir, baseName) {
                    const localBin = path.resolve(baseDir, baseName);
                    if (fs.existsSync(localBin)) return localBin;
                    if (sourceExt.includes(path.extname(baseName))) return;
                    const foundExt = sourceExt.find((ext => fs.existsSync(`${localBin}${ext}`)));
                    return foundExt ? `${localBin}${foundExt}` : void 0;
                }
                this._checkForMissingMandatoryOptions(), this._checkForConflictingOptions();
                let proc, executableFile = subcommand._executableFile || `${this._name}-${subcommand._name}`, executableDir = this._executableDir || "";
                if (this._scriptPath) {
                    let resolvedScriptPath;
                    try {
                        resolvedScriptPath = fs.realpathSync(this._scriptPath);
                    } catch (err) {
                        resolvedScriptPath = this._scriptPath;
                    }
                    executableDir = path.resolve(path.dirname(resolvedScriptPath), executableDir);
                }
                if (executableDir) {
                    let localFile = findFile(executableDir, executableFile);
                    if (!localFile && !subcommand._executableFile && this._scriptPath) {
                        const legacyName = path.basename(this._scriptPath, path.extname(this._scriptPath));
                        legacyName !== this._name && (localFile = findFile(executableDir, `${legacyName}-${subcommand._name}`));
                    }
                    executableFile = localFile || executableFile;
                }
                launchWithNode = sourceExt.includes(path.extname(executableFile)), "win32" !== process.platform ? launchWithNode ? (args.unshift(executableFile), 
                args = incrementNodeInspectorPort(process.execArgv).concat(args), proc = childProcess.spawn(process.argv[0], args, {
                    stdio: "inherit"
                })) : proc = childProcess.spawn(executableFile, args, {
                    stdio: "inherit"
                }) : (args.unshift(executableFile), args = incrementNodeInspectorPort(process.execArgv).concat(args), 
                proc = childProcess.spawn(process.execPath, args, {
                    stdio: "inherit"
                })), proc.killed || [ "SIGUSR1", "SIGUSR2", "SIGTERM", "SIGINT", "SIGHUP" ].forEach((signal => {
                    process.on(signal, (() => {
                        !1 === proc.killed && null === proc.exitCode && proc.kill(signal);
                    }));
                }));
                const exitCallback = this._exitCallback;
                proc.on("close", (code => {
                    code = code ?? 1, exitCallback ? exitCallback(new CommanderError(code, "commander.executeSubCommandAsync", "(close)")) : process.exit(code);
                })), proc.on("error", (err => {
                    if ("ENOENT" === err.code) {
                        const executableDirMessage = executableDir ? `searched for local subcommand relative to directory '${executableDir}'` : "no directory for search for local subcommand, use .executableDir() to supply a custom directory", executableMissing = `'${executableFile}' does not exist\n - if '${subcommand._name}' is not meant to be an executable command, remove description parameter from '.command()' and use '.description()' instead\n - if the default executable name is not suitable, use the executableFile option to supply a custom name or path\n - ${executableDirMessage}`;
                        throw new Error(executableMissing);
                    }
                    if ("EACCES" === err.code) throw new Error(`'${executableFile}' not executable`);
                    if (exitCallback) {
                        const wrappedError = new CommanderError(1, "commander.executeSubCommandAsync", "(error)");
                        wrappedError.nestedError = err, exitCallback(wrappedError);
                    } else process.exit(1);
                })), this.runningCommand = proc;
            }
            _dispatchSubcommand(commandName, operands, unknown) {
                const subCommand = this._findCommand(commandName);
                let promiseChain;
                return subCommand || this.help({
                    error: !0
                }), promiseChain = this._chainOrCallSubCommandHook(promiseChain, subCommand, "preSubcommand"), 
                promiseChain = this._chainOrCall(promiseChain, (() => {
                    if (!subCommand._executableHandler) return subCommand._parseCommand(operands, unknown);
                    this._executeSubCommand(subCommand, operands.concat(unknown));
                })), promiseChain;
            }
            _dispatchHelpCommand(subcommandName) {
                subcommandName || this.help();
                const subCommand = this._findCommand(subcommandName);
                return subCommand && !subCommand._executableHandler && subCommand.help(), this._dispatchSubcommand(subcommandName, [], [ this._getHelpOption()?.long ?? this._getHelpOption()?.short ?? "--help" ]);
            }
            _checkNumberOfArguments() {
                this.registeredArguments.forEach(((arg, i) => {
                    arg.required && null == this.args[i] && this.missingArgument(arg.name());
                })), this.registeredArguments.length > 0 && this.registeredArguments[this.registeredArguments.length - 1].variadic || this.args.length > this.registeredArguments.length && this._excessArguments(this.args);
            }
            _processArguments() {
                const myParseArg = (argument, value, previous) => {
                    let parsedValue = value;
                    if (null !== value && argument.parseArg) {
                        const invalidValueMessage = `error: command-argument value '${value}' is invalid for argument '${argument.name()}'.`;
                        parsedValue = this._callParseArg(argument, value, previous, invalidValueMessage);
                    }
                    return parsedValue;
                };
                this._checkNumberOfArguments();
                const processedArgs = [];
                this.registeredArguments.forEach(((declaredArg, index) => {
                    let value = declaredArg.defaultValue;
                    declaredArg.variadic ? index < this.args.length ? (value = this.args.slice(index), 
                    declaredArg.parseArg && (value = value.reduce(((processed, v) => myParseArg(declaredArg, v, processed)), declaredArg.defaultValue))) : void 0 === value && (value = []) : index < this.args.length && (value = this.args[index], 
                    declaredArg.parseArg && (value = myParseArg(declaredArg, value, declaredArg.defaultValue))), 
                    processedArgs[index] = value;
                })), this.processedArgs = processedArgs;
            }
            _chainOrCall(promise, fn) {
                return promise && promise.then && "function" == typeof promise.then ? promise.then((() => fn())) : fn();
            }
            _chainOrCallHooks(promise, event) {
                let result = promise;
                const hooks = [];
                return this._getCommandAndAncestors().reverse().filter((cmd => void 0 !== cmd._lifeCycleHooks[event])).forEach((hookedCommand => {
                    hookedCommand._lifeCycleHooks[event].forEach((callback => {
                        hooks.push({
                            hookedCommand: hookedCommand,
                            callback: callback
                        });
                    }));
                })), "postAction" === event && hooks.reverse(), hooks.forEach((hookDetail => {
                    result = this._chainOrCall(result, (() => hookDetail.callback(hookDetail.hookedCommand, this)));
                })), result;
            }
            _chainOrCallSubCommandHook(promise, subCommand, event) {
                let result = promise;
                return void 0 !== this._lifeCycleHooks[event] && this._lifeCycleHooks[event].forEach((hook => {
                    result = this._chainOrCall(result, (() => hook(this, subCommand)));
                })), result;
            }
            _parseCommand(operands, unknown) {
                const parsed = this.parseOptions(unknown);
                if (this._parseOptionsEnv(), this._parseOptionsImplied(), operands = operands.concat(parsed.operands), 
                unknown = parsed.unknown, this.args = operands.concat(unknown), operands && this._findCommand(operands[0])) return this._dispatchSubcommand(operands[0], operands.slice(1), unknown);
                if (this._getHelpCommand() && operands[0] === this._getHelpCommand().name()) return this._dispatchHelpCommand(operands[1]);
                if (this._defaultCommandName) return this._outputHelpIfRequested(unknown), this._dispatchSubcommand(this._defaultCommandName, operands, unknown);
                !this.commands.length || 0 !== this.args.length || this._actionHandler || this._defaultCommandName || this.help({
                    error: !0
                }), this._outputHelpIfRequested(parsed.unknown), this._checkForMissingMandatoryOptions(), 
                this._checkForConflictingOptions();
                const checkForUnknownOptions = () => {
                    parsed.unknown.length > 0 && this.unknownOption(parsed.unknown[0]);
                }, commandEvent = `command:${this.name()}`;
                if (this._actionHandler) {
                    let promiseChain;
                    return checkForUnknownOptions(), this._processArguments(), promiseChain = this._chainOrCallHooks(promiseChain, "preAction"), 
                    promiseChain = this._chainOrCall(promiseChain, (() => this._actionHandler(this.processedArgs))), 
                    this.parent && (promiseChain = this._chainOrCall(promiseChain, (() => {
                        this.parent.emit(commandEvent, operands, unknown);
                    }))), promiseChain = this._chainOrCallHooks(promiseChain, "postAction"), promiseChain;
                }
                if (this.parent && this.parent.listenerCount(commandEvent)) checkForUnknownOptions(), 
                this._processArguments(), this.parent.emit(commandEvent, operands, unknown); else if (operands.length) {
                    if (this._findCommand("*")) return this._dispatchSubcommand("*", operands, unknown);
                    this.listenerCount("command:*") ? this.emit("command:*", operands, unknown) : this.commands.length ? this.unknownCommand() : (checkForUnknownOptions(), 
                    this._processArguments());
                } else this.commands.length ? (checkForUnknownOptions(), this.help({
                    error: !0
                })) : (checkForUnknownOptions(), this._processArguments());
            }
            _findCommand(name) {
                if (name) return this.commands.find((cmd => cmd._name === name || cmd._aliases.includes(name)));
            }
            _findOption(arg) {
                return this.options.find((option => option.is(arg)));
            }
            _checkForMissingMandatoryOptions() {
                this._getCommandAndAncestors().forEach((cmd => {
                    cmd.options.forEach((anOption => {
                        anOption.mandatory && void 0 === cmd.getOptionValue(anOption.attributeName()) && cmd.missingMandatoryOptionValue(anOption);
                    }));
                }));
            }
            _checkForConflictingLocalOptions() {
                const definedNonDefaultOptions = this.options.filter((option => {
                    const optionKey = option.attributeName();
                    return void 0 !== this.getOptionValue(optionKey) && "default" !== this.getOptionValueSource(optionKey);
                })), optionsWithConflicting = definedNonDefaultOptions.filter((option => option.conflictsWith.length > 0));
                optionsWithConflicting.forEach((option => {
                    const conflictingAndDefined = definedNonDefaultOptions.find((defined => option.conflictsWith.includes(defined.attributeName())));
                    conflictingAndDefined && this._conflictingOption(option, conflictingAndDefined);
                }));
            }
            _checkForConflictingOptions() {
                this._getCommandAndAncestors().forEach((cmd => {
                    cmd._checkForConflictingLocalOptions();
                }));
            }
            parseOptions(argv) {
                const operands = [], unknown = [];
                let dest = operands;
                const args = argv.slice();
                function maybeOption(arg) {
                    return arg.length > 1 && "-" === arg[0];
                }
                let activeVariadicOption = null;
                for (;args.length; ) {
                    const arg = args.shift();
                    if ("--" === arg) {
                        dest === unknown && dest.push(arg), dest.push(...args);
                        break;
                    }
                    if (!activeVariadicOption || maybeOption(arg)) {
                        if (activeVariadicOption = null, maybeOption(arg)) {
                            const option = this._findOption(arg);
                            if (option) {
                                if (option.required) {
                                    const value = args.shift();
                                    void 0 === value && this.optionMissingArgument(option), this.emit(`option:${option.name()}`, value);
                                } else if (option.optional) {
                                    let value = null;
                                    args.length > 0 && !maybeOption(args[0]) && (value = args.shift()), this.emit(`option:${option.name()}`, value);
                                } else this.emit(`option:${option.name()}`);
                                activeVariadicOption = option.variadic ? option : null;
                                continue;
                            }
                        }
                        if (arg.length > 2 && "-" === arg[0] && "-" !== arg[1]) {
                            const option = this._findOption(`-${arg[1]}`);
                            if (option) {
                                option.required || option.optional && this._combineFlagAndOptionalValue ? this.emit(`option:${option.name()}`, arg.slice(2)) : (this.emit(`option:${option.name()}`), 
                                args.unshift(`-${arg.slice(2)}`));
                                continue;
                            }
                        }
                        if (/^--[^=]+=/.test(arg)) {
                            const index = arg.indexOf("="), option = this._findOption(arg.slice(0, index));
                            if (option && (option.required || option.optional)) {
                                this.emit(`option:${option.name()}`, arg.slice(index + 1));
                                continue;
                            }
                        }
                        if (maybeOption(arg) && (dest = unknown), (this._enablePositionalOptions || this._passThroughOptions) && 0 === operands.length && 0 === unknown.length) {
                            if (this._findCommand(arg)) {
                                operands.push(arg), args.length > 0 && unknown.push(...args);
                                break;
                            }
                            if (this._getHelpCommand() && arg === this._getHelpCommand().name()) {
                                operands.push(arg), args.length > 0 && operands.push(...args);
                                break;
                            }
                            if (this._defaultCommandName) {
                                unknown.push(arg), args.length > 0 && unknown.push(...args);
                                break;
                            }
                        }
                        if (this._passThroughOptions) {
                            dest.push(arg), args.length > 0 && dest.push(...args);
                            break;
                        }
                        dest.push(arg);
                    } else this.emit(`option:${activeVariadicOption.name()}`, arg);
                }
                return {
                    operands: operands,
                    unknown: unknown
                };
            }
            opts() {
                if (this._storeOptionsAsProperties) {
                    const result = {}, len = this.options.length;
                    for (let i = 0; i < len; i++) {
                        const key = this.options[i].attributeName();
                        result[key] = key === this._versionOptionName ? this._version : this[key];
                    }
                    return result;
                }
                return this._optionValues;
            }
            optsWithGlobals() {
                return this._getCommandAndAncestors().reduce(((combinedOptions, cmd) => Object.assign(combinedOptions, cmd.opts())), {});
            }
            error(message, errorOptions) {
                this._outputConfiguration.outputError(`${message}\n`, this._outputConfiguration.writeErr), 
                "string" == typeof this._showHelpAfterError ? this._outputConfiguration.writeErr(`${this._showHelpAfterError}\n`) : this._showHelpAfterError && (this._outputConfiguration.writeErr("\n"), 
                this.outputHelp({
                    error: !0
                }));
                const config = errorOptions || {}, exitCode = config.exitCode || 1, code = config.code || "commander.error";
                this._exit(exitCode, code, message);
            }
            _parseOptionsEnv() {
                this.options.forEach((option => {
                    if (option.envVar && option.envVar in process.env) {
                        const optionKey = option.attributeName();
                        (void 0 === this.getOptionValue(optionKey) || [ "default", "config", "env" ].includes(this.getOptionValueSource(optionKey))) && (option.required || option.optional ? this.emit(`optionEnv:${option.name()}`, process.env[option.envVar]) : this.emit(`optionEnv:${option.name()}`));
                    }
                }));
            }
            _parseOptionsImplied() {
                const dualHelper = new DualOptions(this.options), hasCustomOptionValue = optionKey => void 0 !== this.getOptionValue(optionKey) && ![ "default", "implied" ].includes(this.getOptionValueSource(optionKey));
                this.options.filter((option => void 0 !== option.implied && hasCustomOptionValue(option.attributeName()) && dualHelper.valueFromOption(this.getOptionValue(option.attributeName()), option))).forEach((option => {
                    Object.keys(option.implied).filter((impliedKey => !hasCustomOptionValue(impliedKey))).forEach((impliedKey => {
                        this.setOptionValueWithSource(impliedKey, option.implied[impliedKey], "implied");
                    }));
                }));
            }
            missingArgument(name) {
                const message = `error: missing required argument '${name}'`;
                this.error(message, {
                    code: "commander.missingArgument"
                });
            }
            optionMissingArgument(option) {
                const message = `error: option '${option.flags}' argument missing`;
                this.error(message, {
                    code: "commander.optionMissingArgument"
                });
            }
            missingMandatoryOptionValue(option) {
                const message = `error: required option '${option.flags}' not specified`;
                this.error(message, {
                    code: "commander.missingMandatoryOptionValue"
                });
            }
            _conflictingOption(option, conflictingOption) {
                const findBestOptionFromValue = option => {
                    const optionKey = option.attributeName(), optionValue = this.getOptionValue(optionKey), negativeOption = this.options.find((target => target.negate && optionKey === target.attributeName())), positiveOption = this.options.find((target => !target.negate && optionKey === target.attributeName()));
                    return negativeOption && (void 0 === negativeOption.presetArg && !1 === optionValue || void 0 !== negativeOption.presetArg && optionValue === negativeOption.presetArg) ? negativeOption : positiveOption || option;
                }, getErrorMessage = option => {
                    const bestOption = findBestOptionFromValue(option), optionKey = bestOption.attributeName();
                    return "env" === this.getOptionValueSource(optionKey) ? `environment variable '${bestOption.envVar}'` : `option '${bestOption.flags}'`;
                }, message = `error: ${getErrorMessage(option)} cannot be used with ${getErrorMessage(conflictingOption)}`;
                this.error(message, {
                    code: "commander.conflictingOption"
                });
            }
            unknownOption(flag) {
                if (this._allowUnknownOption) return;
                let suggestion = "";
                if (flag.startsWith("--") && this._showSuggestionAfterError) {
                    let candidateFlags = [], command = this;
                    do {
                        const moreFlags = command.createHelp().visibleOptions(command).filter((option => option.long)).map((option => option.long));
                        candidateFlags = candidateFlags.concat(moreFlags), command = command.parent;
                    } while (command && !command._enablePositionalOptions);
                    suggestion = suggestSimilar(flag, candidateFlags);
                }
                const message = `error: unknown option '${flag}'${suggestion}`;
                this.error(message, {
                    code: "commander.unknownOption"
                });
            }
            _excessArguments(receivedArgs) {
                if (this._allowExcessArguments) return;
                const expected = this.registeredArguments.length, s = 1 === expected ? "" : "s", message = `error: too many arguments${this.parent ? ` for '${this.name()}'` : ""}. Expected ${expected} argument${s} but got ${receivedArgs.length}.`;
                this.error(message, {
                    code: "commander.excessArguments"
                });
            }
            unknownCommand() {
                const unknownName = this.args[0];
                let suggestion = "";
                if (this._showSuggestionAfterError) {
                    const candidateNames = [];
                    this.createHelp().visibleCommands(this).forEach((command => {
                        candidateNames.push(command.name()), command.alias() && candidateNames.push(command.alias());
                    })), suggestion = suggestSimilar(unknownName, candidateNames);
                }
                const message = `error: unknown command '${unknownName}'${suggestion}`;
                this.error(message, {
                    code: "commander.unknownCommand"
                });
            }
            version(str, flags, description) {
                if (void 0 === str) return this._version;
                this._version = str, flags = flags || "-V, --version", description = description || "output the version number";
                const versionOption = this.createOption(flags, description);
                return this._versionOptionName = versionOption.attributeName(), this._registerOption(versionOption), 
                this.on("option:" + versionOption.name(), (() => {
                    this._outputConfiguration.writeOut(`${str}\n`), this._exit(0, "commander.version", str);
                })), this;
            }
            description(str, argsDescription) {
                return void 0 === str && void 0 === argsDescription ? this._description : (this._description = str, 
                argsDescription && (this._argsDescription = argsDescription), this);
            }
            summary(str) {
                return void 0 === str ? this._summary : (this._summary = str, this);
            }
            alias(alias) {
                if (void 0 === alias) return this._aliases[0];
                let command = this;
                if (0 !== this.commands.length && this.commands[this.commands.length - 1]._executableHandler && (command = this.commands[this.commands.length - 1]), 
                alias === command._name) throw new Error("Command alias can't be the same as its name");
                const matchingCommand = this.parent?._findCommand(alias);
                if (matchingCommand) {
                    const existingCmd = [ matchingCommand.name() ].concat(matchingCommand.aliases()).join("|");
                    throw new Error(`cannot add alias '${alias}' to command '${this.name()}' as already have command '${existingCmd}'`);
                }
                return command._aliases.push(alias), this;
            }
            aliases(aliases) {
                return void 0 === aliases ? this._aliases : (aliases.forEach((alias => this.alias(alias))), 
                this);
            }
            usage(str) {
                if (void 0 === str) {
                    if (this._usage) return this._usage;
                    const args = this.registeredArguments.map((arg => humanReadableArgName(arg)));
                    return [].concat(this.options.length || null !== this._helpOption ? "[options]" : [], this.commands.length ? "[command]" : [], this.registeredArguments.length ? args : []).join(" ");
                }
                return this._usage = str, this;
            }
            name(str) {
                return void 0 === str ? this._name : (this._name = str, this);
            }
            nameFromFilename(filename) {
                return this._name = path.basename(filename, path.extname(filename)), this;
            }
            executableDir(path) {
                return void 0 === path ? this._executableDir : (this._executableDir = path, this);
            }
            helpInformation(contextOptions) {
                const helper = this.createHelp();
                return void 0 === helper.helpWidth && (helper.helpWidth = contextOptions && contextOptions.error ? this._outputConfiguration.getErrHelpWidth() : this._outputConfiguration.getOutHelpWidth()), 
                helper.formatHelp(this, helper);
            }
            _getHelpContext(contextOptions) {
                const context = {
                    error: !!(contextOptions = contextOptions || {}).error
                };
                let write;
                return write = context.error ? arg => this._outputConfiguration.writeErr(arg) : arg => this._outputConfiguration.writeOut(arg), 
                context.write = contextOptions.write || write, context.command = this, context;
            }
            outputHelp(contextOptions) {
                let deprecatedCallback;
                "function" == typeof contextOptions && (deprecatedCallback = contextOptions, contextOptions = void 0);
                const context = this._getHelpContext(contextOptions);
                this._getCommandAndAncestors().reverse().forEach((command => command.emit("beforeAllHelp", context))), 
                this.emit("beforeHelp", context);
                let helpInformation = this.helpInformation(context);
                if (deprecatedCallback && (helpInformation = deprecatedCallback(helpInformation), 
                "string" != typeof helpInformation && !Buffer.isBuffer(helpInformation))) throw new Error("outputHelp callback must return a string or a Buffer");
                context.write(helpInformation), this._getHelpOption()?.long && this.emit(this._getHelpOption().long), 
                this.emit("afterHelp", context), this._getCommandAndAncestors().forEach((command => command.emit("afterAllHelp", context)));
            }
            helpOption(flags, description) {
                return "boolean" == typeof flags ? (this._helpOption = flags ? this._helpOption ?? void 0 : null, 
                this) : (flags = flags ?? "-h, --help", description = description ?? "display help for command", 
                this._helpOption = this.createOption(flags, description), this);
            }
            _getHelpOption() {
                return void 0 === this._helpOption && this.helpOption(void 0, void 0), this._helpOption;
            }
            addHelpOption(option) {
                return this._helpOption = option, this;
            }
            help(contextOptions) {
                this.outputHelp(contextOptions);
                let exitCode = process.exitCode || 0;
                0 === exitCode && contextOptions && "function" != typeof contextOptions && contextOptions.error && (exitCode = 1), 
                this._exit(exitCode, "commander.help", "(outputHelp)");
            }
            addHelpText(position, text) {
                const allowedValues = [ "beforeAll", "before", "after", "afterAll" ];
                if (!allowedValues.includes(position)) throw new Error(`Unexpected value for position to addHelpText.\nExpecting one of '${allowedValues.join("', '")}'`);
                const helpEvent = `${position}Help`;
                return this.on(helpEvent, (context => {
                    let helpStr;
                    helpStr = "function" == typeof text ? text({
                        error: context.error,
                        command: context.command
                    }) : text, helpStr && context.write(`${helpStr}\n`);
                })), this;
            }
            _outputHelpIfRequested(args) {
                const helpOption = this._getHelpOption();
                helpOption && args.find((arg => helpOption.is(arg))) && (this.outputHelp(), this._exit(0, "commander.helpDisplayed", "(outputHelp)"));
            }
        }
        function incrementNodeInspectorPort(args) {
            return args.map((arg => {
                if (!arg.startsWith("--inspect")) return arg;
                let debugOption, match, debugHost = "127.0.0.1", debugPort = "9229";
                return null !== (match = arg.match(/^(--inspect(-brk)?)$/)) ? debugOption = match[1] : null !== (match = arg.match(/^(--inspect(-brk|-port)?)=([^:]+)$/)) ? (debugOption = match[1], 
                /^\d+$/.test(match[3]) ? debugPort = match[3] : debugHost = match[3]) : null !== (match = arg.match(/^(--inspect(-brk|-port)?)=([^:]+):(\d+)$/)) && (debugOption = match[1], 
                debugHost = match[3], debugPort = match[4]), debugOption && "0" !== debugPort ? `${debugOption}=${debugHost}:${parseInt(debugPort) + 1}` : arg;
            }));
        }
        return command.Command = Command, command;
    }(), {CommanderError: CommanderError, InvalidArgumentError: InvalidArgumentError} = requireError(), {Help: Help} = requireHelp(), {Option: Option} = requireOption();
    return commander.program = new Command, commander.createCommand = name => new Command(name), 
    commander.createOption = (flags, description) => new Option(flags, description), 
    commander.createArgument = (name, description) => new Argument(name, description), 
    commander.Command = Command, commander.Option = Option, commander.Argument = Argument, 
    commander.Help = Help, commander.CommanderError = CommanderError, commander.InvalidArgumentError = InvalidArgumentError, 
    commander.InvalidOptionArgumentError = InvalidArgumentError, commander;
}

var extraTypingsExports = (hasRequiredExtraTypings || (hasRequiredExtraTypings = 1, 
function(module, exports) {
    const commander = requireCommander();
    (exports = module.exports = {}).program = new commander.Command, exports.Argument = commander.Argument, 
    exports.Command = commander.Command, exports.CommanderError = commander.CommanderError, 
    exports.Help = commander.Help, exports.InvalidArgumentError = commander.InvalidArgumentError, 
    exports.InvalidOptionArgumentError = commander.InvalidArgumentError, exports.Option = commander.Option, 
    exports.createCommand = name => new commander.Command(name), exports.createOption = (flags, description) => new commander.Option(flags, description), 
    exports.createArgument = (name, description) => new commander.Argument(name, description);
}(extraTypings, extraTypings.exports)), extraTypings.exports), extraTypingsCommander = getDefaultExportFromCjs(extraTypingsExports);

const {program: program, createCommand: createCommand, createArgument: createArgument, createOption: createOption, CommanderError: CommanderError, InvalidArgumentError: InvalidArgumentError, InvalidOptionArgumentError: InvalidOptionArgumentError, Command: Command, Argument: Argument, Option: Option, Help: Help} = extraTypingsCommander, cliOutputConfig = {
    writeErr(str) {
        console.log(str.trim() + "\n");
    },
    writeOut(str) {
        console.log(str.trim() + "\n");
    }
};

var packageJson_name = "solana", packageJson_version = "3.0.0";

function cliProgramRoot() {
    const app = {
        version: packageJson_version,
        name: packageJson_name
    };
    return (new Command).name("npx solana").version(app.version, "--version", "output the version number of this tool").configureOutput(cliOutputConfig);
}

var __defProp = Object.defineProperty, __defProps = Object.defineProperties, __getOwnPropDescs = Object.getOwnPropertyDescriptors, __getOwnPropSymbols = Object.getOwnPropertySymbols, __hasOwnProp = Object.prototype.hasOwnProperty, __propIsEnum = Object.prototype.propertyIsEnumerable, __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, {
    enumerable: !0,
    configurable: !0,
    writable: !0,
    value: value
}) : obj[key] = value, __spreadValues = (a, b) => {
    for (var prop in b || (b = {})) __hasOwnProp.call(b, prop) && __defNormalProp(a, prop, b[prop]);
    if (__getOwnPropSymbols) for (var prop of __getOwnPropSymbols(b)) __propIsEnum.call(b, prop) && __defNormalProp(a, prop, b[prop]);
    return a;
}, __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));

var src_default = function(cmd, opts) {
    const executable = Array.isArray(cmd) ? cmd.join(";") : cmd, options = __spreadProps(__spreadValues({}, opts), {
        stdio: "pipe",
        cwd: process.cwd()
    }), {platform: platform} = process;
    try {
        const cmd2 = "win32" === platform ? "cmd" : "sh", arg = "win32" === platform ? "/C" : "-c", child = childProcess.spawn(cmd2, [ arg, executable ], options);
        return new Promise((resolve => {
            const stdoutList = [], stderrList = [];
            child.stdout && child.stdout.on("data", (data => {
                if (Buffer.isBuffer(data)) return stdoutList.push(data.toString());
                stdoutList.push(data);
            })), child.stderr && child.stderr.on("data", (data => {
                if (Buffer.isBuffer(data)) return stderrList.push(data.toString());
                stderrList.push(JSON.stringify(data));
            }));
            const getDefaultResult = () => {
                const stderr = stderrList.join("\n");
                return {
                    stdout: stdoutList.join("\n"),
                    stderr: stderr,
                    cmd: executable
                };
            };
            child.on("error", (error => resolve(__spreadProps(__spreadValues({}, getDefaultResult()), {
                error: error
            })))), child.on("close", (code => resolve(__spreadProps(__spreadValues({}, getDefaultResult()), {
                code: code
            }))));
        }));
    } catch (error) {
        return Promise.reject(error);
    }
};

const TOOL_CONFIG = {
    rust: {
        pathSource: "$HOME/.cargo/env",
        version: "rustc --version"
    },
    solana: {
        dependencies: [ "rust" ],
        pathSource: "$HOME/.local/share/solana/install/active_release/bin",
        version: "solana --version"
    },
    avm: {
        dependencies: [ "rust" ],
        version: "avm --version"
    },
    anchor: {
        dependencies: [ "avm" ],
        version: "anchor --version"
    },
    yarn: {
        version: "yarn --version"
    },
    mucho: {
        version: "mucho --version"
    },
    trident: {
        dependencies: [ "rust" ],
        version: "cargo install --list | grep trident-cli"
    },
    zest: {
        dependencies: [ "rust" ],
        version: "cargo install --list | grep zest"
    },
    verify: {
        dependencies: [ "rust" ],
        version: "solana-verify --version"
    }
};

var PathSourceStatus;

!function(PathSourceStatus) {
    PathSourceStatus[PathSourceStatus.SUCCESS = 0] = "SUCCESS", PathSourceStatus[PathSourceStatus.FAILED = 1] = "FAILED", 
    PathSourceStatus[PathSourceStatus.OUTPUT_MISMATCH = 2] = "OUTPUT_MISMATCH", PathSourceStatus[PathSourceStatus.MISSING_PATH = 3] = "MISSING_PATH";
}(PathSourceStatus || (PathSourceStatus = {}));

const VERSION_REGEX = /(?:[\w-]+\s+)?(\d+\.\d+\.?\d+?)/;

async function installedToolVersion(name) {
    let command = "";
    if (Object.prototype.hasOwnProperty.call(TOOL_CONFIG, name) && (command = TOOL_CONFIG[name].version, 
    TOOL_CONFIG[name].pathSource && (command = `export PATH="${TOOL_CONFIG[name].pathSource}:$PATH" && ${command}`)), 
    !command) return !1;
    const res = await checkCommand(command);
    return res && VERSION_REGEX.exec(res)[1] || res;
}

async function checkCommand(cmd, errorOptions = null) {
    try {
        const {stdout: stdout} = await src_default(cmd);
        if (stdout) return stdout.trim();
        if (errorOptions) throw "Command not found";
        return !1;
    } catch (err) {
        if (errorOptions) if ("function" == typeof errorOptions?.onError) {
            const res = await errorOptions.onError();
            if (errorOptions.doubleCheck) return await checkCommand(cmd, {
                exit: errorOptions.exit
            });
            !res && errorOptions?.exit && process.exit(1);
        } else warnMessage(errorOptions.message || `Unable to execute command: ${cmd.split(" ")[0]}`), 
        errorOptions.exit && process.exit(1);
        return !1;
    }
}

function detectOperatingSystem() {
    switch (os.platform()) {
      case "darwin":
        return "mac";

      case "win32":
        return "windows";

      case "linux":
        return "linux";

      default:
        return "unknown";
    }
}

function appendPathToRCFiles(newPath, name = void 0) {
    const homeDir = process.env.HOME || process.env.USERPROFILE;
    if (!homeDir) return void console.error("[appendPathToRCFiles]", "Unable to find the user home directory.");
    const rcFiles = [ ".bashrc", ".zshrc" ].map((file => path.join(homeDir, file)));
    let exportLine = `export PATH=${newPath}:$PATH\n`;
    if (name) {
        const exportName = `${name.toUpperCase()}_HOME`;
        exportLine = `\n# ${name}\nexport ${exportName}="${newPath}"\ncase ":$PATH:" in\n  *":$${exportName}:"*) ;;\n  *) export PATH="$${exportName}:$PATH" ;;\nesac\n# ${name} end\n\n`;
    }
    rcFiles.forEach((rcFile => {
        if (fs.existsSync(rcFile)) {
            fs.readFileSync(rcFile, "utf-8").includes(newPath) || fs.appendFileSync(rcFile, exportLine);
        }
    }));
}

const wrapAnsi16 = (offset = 0) => code => `[${code + offset}m`, wrapAnsi256 = (offset = 0) => code => `[${38 + offset};5;${code}m`, wrapAnsi16m = (offset = 0) => (red, green, blue) => `[${38 + offset};2;${red};${green};${blue}m`, styles$1 = {
    modifier: {
        reset: [ 0, 0 ],
        bold: [ 1, 22 ],
        dim: [ 2, 22 ],
        italic: [ 3, 23 ],
        underline: [ 4, 24 ],
        overline: [ 53, 55 ],
        inverse: [ 7, 27 ],
        hidden: [ 8, 28 ],
        strikethrough: [ 9, 29 ]
    },
    color: {
        black: [ 30, 39 ],
        red: [ 31, 39 ],
        green: [ 32, 39 ],
        yellow: [ 33, 39 ],
        blue: [ 34, 39 ],
        magenta: [ 35, 39 ],
        cyan: [ 36, 39 ],
        white: [ 37, 39 ],
        blackBright: [ 90, 39 ],
        gray: [ 90, 39 ],
        grey: [ 90, 39 ],
        redBright: [ 91, 39 ],
        greenBright: [ 92, 39 ],
        yellowBright: [ 93, 39 ],
        blueBright: [ 94, 39 ],
        magentaBright: [ 95, 39 ],
        cyanBright: [ 96, 39 ],
        whiteBright: [ 97, 39 ]
    },
    bgColor: {
        bgBlack: [ 40, 49 ],
        bgRed: [ 41, 49 ],
        bgGreen: [ 42, 49 ],
        bgYellow: [ 43, 49 ],
        bgBlue: [ 44, 49 ],
        bgMagenta: [ 45, 49 ],
        bgCyan: [ 46, 49 ],
        bgWhite: [ 47, 49 ],
        bgBlackBright: [ 100, 49 ],
        bgGray: [ 100, 49 ],
        bgGrey: [ 100, 49 ],
        bgRedBright: [ 101, 49 ],
        bgGreenBright: [ 102, 49 ],
        bgYellowBright: [ 103, 49 ],
        bgBlueBright: [ 104, 49 ],
        bgMagentaBright: [ 105, 49 ],
        bgCyanBright: [ 106, 49 ],
        bgWhiteBright: [ 107, 49 ]
    }
};

Object.keys(styles$1.modifier);

Object.keys(styles$1.color), Object.keys(styles$1.bgColor);

const ansiStyles = function() {
    const codes = new Map;
    for (const [groupName, group] of Object.entries(styles$1)) {
        for (const [styleName, style] of Object.entries(group)) styles$1[styleName] = {
            open: `[${style[0]}m`,
            close: `[${style[1]}m`
        }, group[styleName] = styles$1[styleName], codes.set(style[0], style[1]);
        Object.defineProperty(styles$1, groupName, {
            value: group,
            enumerable: !1
        });
    }
    return Object.defineProperty(styles$1, "codes", {
        value: codes,
        enumerable: !1
    }), styles$1.color.close = "[39m", styles$1.bgColor.close = "[49m", styles$1.color.ansi = wrapAnsi16(), 
    styles$1.color.ansi256 = wrapAnsi256(), styles$1.color.ansi16m = wrapAnsi16m(), 
    styles$1.bgColor.ansi = wrapAnsi16(10), styles$1.bgColor.ansi256 = wrapAnsi256(10), 
    styles$1.bgColor.ansi16m = wrapAnsi16m(10), Object.defineProperties(styles$1, {
        rgbToAnsi256: {
            value: (red, green, blue) => red === green && green === blue ? red < 8 ? 16 : red > 248 ? 231 : Math.round((red - 8) / 247 * 24) + 232 : 16 + 36 * Math.round(red / 255 * 5) + 6 * Math.round(green / 255 * 5) + Math.round(blue / 255 * 5),
            enumerable: !1
        },
        hexToRgb: {
            value(hex) {
                const matches = /[a-f\d]{6}|[a-f\d]{3}/i.exec(hex.toString(16));
                if (!matches) return [ 0, 0, 0 ];
                let [colorString] = matches;
                3 === colorString.length && (colorString = [ ...colorString ].map((character => character + character)).join(""));
                const integer = Number.parseInt(colorString, 16);
                return [ integer >> 16 & 255, integer >> 8 & 255, 255 & integer ];
            },
            enumerable: !1
        },
        hexToAnsi256: {
            value: hex => styles$1.rgbToAnsi256(...styles$1.hexToRgb(hex)),
            enumerable: !1
        },
        ansi256ToAnsi: {
            value(code) {
                if (code < 8) return 30 + code;
                if (code < 16) return code - 8 + 90;
                let red, green, blue;
                if (code >= 232) red = (10 * (code - 232) + 8) / 255, green = red, blue = red; else {
                    const remainder = (code -= 16) % 36;
                    red = Math.floor(code / 36) / 5, green = Math.floor(remainder / 6) / 5, blue = remainder % 6 / 5;
                }
                const value = 2 * Math.max(red, green, blue);
                if (0 === value) return 30;
                let result = 30 + (Math.round(blue) << 2 | Math.round(green) << 1 | Math.round(red));
                return 2 === value && (result += 60), result;
            },
            enumerable: !1
        },
        rgbToAnsi: {
            value: (red, green, blue) => styles$1.ansi256ToAnsi(styles$1.rgbToAnsi256(red, green, blue)),
            enumerable: !1
        },
        hexToAnsi: {
            value: hex => styles$1.ansi256ToAnsi(styles$1.hexToAnsi256(hex)),
            enumerable: !1
        }
    }), styles$1;
}();

function hasFlag(flag, argv = (globalThis.Deno ? globalThis.Deno.args : process$2.argv)) {
    const prefix = flag.startsWith("-") ? "" : 1 === flag.length ? "-" : "--", position = argv.indexOf(prefix + flag), terminatorPosition = argv.indexOf("--");
    return -1 !== position && (-1 === terminatorPosition || position < terminatorPosition);
}

const {env: env} = process$2;

let flagForceColor;

function _supportsColor(haveStream, {streamIsTTY: streamIsTTY, sniffFlags: sniffFlags = !0} = {}) {
    const noFlagForceColor = function() {
        if ("FORCE_COLOR" in env) return "true" === env.FORCE_COLOR ? 1 : "false" === env.FORCE_COLOR ? 0 : 0 === env.FORCE_COLOR.length ? 1 : Math.min(Number.parseInt(env.FORCE_COLOR, 10), 3);
    }();
    void 0 !== noFlagForceColor && (flagForceColor = noFlagForceColor);
    const forceColor = sniffFlags ? flagForceColor : noFlagForceColor;
    if (0 === forceColor) return 0;
    if (sniffFlags) {
        if (hasFlag("color=16m") || hasFlag("color=full") || hasFlag("color=truecolor")) return 3;
        if (hasFlag("color=256")) return 2;
    }
    if ("TF_BUILD" in env && "AGENT_NAME" in env) return 1;
    if (haveStream && !streamIsTTY && void 0 === forceColor) return 0;
    const min = forceColor || 0;
    if ("dumb" === env.TERM) return min;
    if ("win32" === process$2.platform) {
        const osRelease = os.release().split(".");
        return Number(osRelease[0]) >= 10 && Number(osRelease[2]) >= 10586 ? Number(osRelease[2]) >= 14931 ? 3 : 2 : 1;
    }
    if ("CI" in env) return "GITHUB_ACTIONS" in env || "GITEA_ACTIONS" in env ? 3 : [ "TRAVIS", "CIRCLECI", "APPVEYOR", "GITLAB_CI", "BUILDKITE", "DRONE" ].some((sign => sign in env)) || "codeship" === env.CI_NAME ? 1 : min;
    if ("TEAMCITY_VERSION" in env) return /^(9\.(0*[1-9]\d*)\.|\d{2,}\.)/.test(env.TEAMCITY_VERSION) ? 1 : 0;
    if ("truecolor" === env.COLORTERM) return 3;
    if ("xterm-kitty" === env.TERM) return 3;
    if ("TERM_PROGRAM" in env) {
        const version = Number.parseInt((env.TERM_PROGRAM_VERSION || "").split(".")[0], 10);
        switch (env.TERM_PROGRAM) {
          case "iTerm.app":
            return version >= 3 ? 3 : 2;

          case "Apple_Terminal":
            return 2;
        }
    }
    return /-256(color)?$/i.test(env.TERM) ? 2 : /^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env.TERM) || "COLORTERM" in env ? 1 : min;
}

function createSupportsColor(stream, options = {}) {
    return function(level) {
        return 0 !== level && {
            level: level,
            hasBasic: !0,
            has256: level >= 2,
            has16m: level >= 3
        };
    }(_supportsColor(stream, {
        streamIsTTY: stream && stream.isTTY,
        ...options
    }));
}

hasFlag("no-color") || hasFlag("no-colors") || hasFlag("color=false") || hasFlag("color=never") ? flagForceColor = 0 : (hasFlag("color") || hasFlag("colors") || hasFlag("color=true") || hasFlag("color=always")) && (flagForceColor = 1);

const supportsColor = {
    stdout: createSupportsColor({
        isTTY: tty.isatty(1)
    }),
    stderr: createSupportsColor({
        isTTY: tty.isatty(2)
    })
};

function stringReplaceAll(string, substring, replacer) {
    let index = string.indexOf(substring);
    if (-1 === index) return string;
    const substringLength = substring.length;
    let endIndex = 0, returnValue = "";
    do {
        returnValue += string.slice(endIndex, index) + substring + replacer, endIndex = index + substringLength, 
        index = string.indexOf(substring, endIndex);
    } while (-1 !== index);
    return returnValue += string.slice(endIndex), returnValue;
}

const {stdout: stdoutColor, stderr: stderrColor} = supportsColor, GENERATOR = Symbol("GENERATOR"), STYLER = Symbol("STYLER"), IS_EMPTY = Symbol("IS_EMPTY"), levelMapping = [ "ansi", "ansi", "ansi256", "ansi16m" ], styles = Object.create(null), chalkFactory = options => {
    const chalk = (...strings) => strings.join(" ");
    return ((object, options = {}) => {
        if (options.level && !(Number.isInteger(options.level) && options.level >= 0 && options.level <= 3)) throw new Error("The `level` option should be an integer from 0 to 3");
        const colorLevel = stdoutColor ? stdoutColor.level : 0;
        object.level = void 0 === options.level ? colorLevel : options.level;
    })(chalk, options), Object.setPrototypeOf(chalk, createChalk.prototype), chalk;
};

function createChalk(options) {
    return chalkFactory(options);
}

Object.setPrototypeOf(createChalk.prototype, Function.prototype);

for (const [styleName, style] of Object.entries(ansiStyles)) styles[styleName] = {
    get() {
        const builder = createBuilder(this, createStyler(style.open, style.close, this[STYLER]), this[IS_EMPTY]);
        return Object.defineProperty(this, styleName, {
            value: builder
        }), builder;
    }
};

styles.visible = {
    get() {
        const builder = createBuilder(this, this[STYLER], !0);
        return Object.defineProperty(this, "visible", {
            value: builder
        }), builder;
    }
};

const getModelAnsi = (model, level, type, ...arguments_) => "rgb" === model ? "ansi16m" === level ? ansiStyles[type].ansi16m(...arguments_) : "ansi256" === level ? ansiStyles[type].ansi256(ansiStyles.rgbToAnsi256(...arguments_)) : ansiStyles[type].ansi(ansiStyles.rgbToAnsi(...arguments_)) : "hex" === model ? getModelAnsi("rgb", level, type, ...ansiStyles.hexToRgb(...arguments_)) : ansiStyles[type][model](...arguments_), usedModels = [ "rgb", "hex", "ansi256" ];

for (const model of usedModels) {
    styles[model] = {
        get() {
            const {level: level} = this;
            return function(...arguments_) {
                const styler = createStyler(getModelAnsi(model, levelMapping[level], "color", ...arguments_), ansiStyles.color.close, this[STYLER]);
                return createBuilder(this, styler, this[IS_EMPTY]);
            };
        }
    };
    styles["bg" + model[0].toUpperCase() + model.slice(1)] = {
        get() {
            const {level: level} = this;
            return function(...arguments_) {
                const styler = createStyler(getModelAnsi(model, levelMapping[level], "bgColor", ...arguments_), ansiStyles.bgColor.close, this[STYLER]);
                return createBuilder(this, styler, this[IS_EMPTY]);
            };
        }
    };
}

const proto = Object.defineProperties((() => {}), {
    ...styles,
    level: {
        enumerable: !0,
        get() {
            return this[GENERATOR].level;
        },
        set(level) {
            this[GENERATOR].level = level;
        }
    }
}), createStyler = (open, close, parent) => {
    let openAll, closeAll;
    return void 0 === parent ? (openAll = open, closeAll = close) : (openAll = parent.openAll + open, 
    closeAll = close + parent.closeAll), {
        open: open,
        close: close,
        openAll: openAll,
        closeAll: closeAll,
        parent: parent
    };
}, createBuilder = (self, _styler, _isEmpty) => {
    const builder = (...arguments_) => applyStyle(builder, 1 === arguments_.length ? "" + arguments_[0] : arguments_.join(" "));
    return Object.setPrototypeOf(builder, proto), builder[GENERATOR] = self, builder[STYLER] = _styler, 
    builder[IS_EMPTY] = _isEmpty, builder;
}, applyStyle = (self, string) => {
    if (self.level <= 0 || !string) return self[IS_EMPTY] ? "" : string;
    let styler = self[STYLER];
    if (void 0 === styler) return string;
    const {openAll: openAll, closeAll: closeAll} = styler;
    if (string.includes("")) for (;void 0 !== styler; ) string = stringReplaceAll(string, styler.close, styler.open), 
    styler = styler.parent;
    const lfIndex = string.indexOf("\n");
    return -1 !== lfIndex && (string = function(string, prefix, postfix, index) {
        let endIndex = 0, returnValue = "";
        do {
            const gotCR = "\r" === string[index - 1];
            returnValue += string.slice(endIndex, gotCR ? index - 1 : index) + prefix + (gotCR ? "\r\n" : "\n") + postfix, 
            endIndex = index + 1, index = string.indexOf("\n", endIndex);
        } while (-1 !== index);
        return returnValue += string.slice(endIndex), returnValue;
    }(string, closeAll, openAll, lfIndex)), openAll + string + closeAll;
};

Object.defineProperties(createChalk.prototype, styles);

const chalk = createChalk();

createChalk({
    level: stderrColor ? stderrColor.level : 0
});

const copyProperty = (to, from, property, ignoreNonConfigurable) => {
    if ("length" === property || "prototype" === property) return;
    if ("arguments" === property || "caller" === property) return;
    const toDescriptor = Object.getOwnPropertyDescriptor(to, property), fromDescriptor = Object.getOwnPropertyDescriptor(from, property);
    !canCopyProperty(toDescriptor, fromDescriptor) && ignoreNonConfigurable || Object.defineProperty(to, property, fromDescriptor);
}, canCopyProperty = function(toDescriptor, fromDescriptor) {
    return void 0 === toDescriptor || toDescriptor.configurable || toDescriptor.writable === fromDescriptor.writable && toDescriptor.enumerable === fromDescriptor.enumerable && toDescriptor.configurable === fromDescriptor.configurable && (toDescriptor.writable || toDescriptor.value === fromDescriptor.value);
}, wrappedToString = (withName, fromBody) => `/* Wrapped ${withName}*/\n${fromBody}`, toStringDescriptor = Object.getOwnPropertyDescriptor(Function.prototype, "toString"), toStringName = Object.getOwnPropertyDescriptor(Function.prototype.toString, "name");

function mimicFunction(to, from, {ignoreNonConfigurable: ignoreNonConfigurable = !1} = {}) {
    const {name: name} = to;
    for (const property of Reflect.ownKeys(from)) copyProperty(to, from, property, ignoreNonConfigurable);
    return ((to, from) => {
        const fromPrototype = Object.getPrototypeOf(from);
        fromPrototype !== Object.getPrototypeOf(to) && Object.setPrototypeOf(to, fromPrototype);
    })(to, from), ((to, from, name) => {
        const withName = "" === name ? "" : `with ${name.trim()}() `, newToString = wrappedToString.bind(null, withName, from.toString());
        Object.defineProperty(newToString, "name", toStringName);
        const {writable: writable, enumerable: enumerable, configurable: configurable} = toStringDescriptor;
        Object.defineProperty(to, "toString", {
            value: newToString,
            writable: writable,
            enumerable: enumerable,
            configurable: configurable
        });
    })(to, from, name), to;
}

const calledFunctions = new WeakMap, onetime = (function_, options = {}) => {
    if ("function" != typeof function_) throw new TypeError("Expected a function");
    let returnValue, callCount = 0;
    const functionName = function_.displayName || function_.name || "<anonymous>", onetime = function(...arguments_) {
        if (calledFunctions.set(onetime, ++callCount), 1 === callCount) returnValue = function_.apply(this, arguments_), 
        function_ = void 0; else if (!0 === options.throw) throw new Error(`Function \`${functionName}\` can only be called once`);
        return returnValue;
    };
    return mimicFunction(onetime, function_), calledFunctions.set(onetime, callCount), 
    onetime;
};

onetime.callCount = function_ => {
    if (!calledFunctions.has(function_)) throw new Error(`The given function \`${function_.name}\` is not wrapped by the \`onetime\` package`);
    return calledFunctions.get(function_);
};

const signals = [];

signals.push("SIGHUP", "SIGINT", "SIGTERM"), "win32" !== process.platform && signals.push("SIGALRM", "SIGABRT", "SIGVTALRM", "SIGXCPU", "SIGXFSZ", "SIGUSR2", "SIGTRAP", "SIGSYS", "SIGQUIT", "SIGIOT"), 
"linux" === process.platform && signals.push("SIGIO", "SIGPOLL", "SIGPWR", "SIGSTKFLT");

const processOk = process => !!process && "object" == typeof process && "function" == typeof process.removeListener && "function" == typeof process.emit && "function" == typeof process.reallyExit && "function" == typeof process.listeners && "function" == typeof process.kill && "number" == typeof process.pid && "function" == typeof process.on, kExitEmitter = Symbol.for("signal-exit emitter"), global = globalThis, ObjectDefineProperty = Object.defineProperty.bind(Object);

class Emitter {
    emitted={
        afterExit: !1,
        exit: !1
    };
    listeners={
        afterExit: [],
        exit: []
    };
    count=0;
    id=Math.random();
    constructor() {
        if (global[kExitEmitter]) return global[kExitEmitter];
        ObjectDefineProperty(global, kExitEmitter, {
            value: this,
            writable: !1,
            enumerable: !1,
            configurable: !1
        });
    }
    on(ev, fn) {
        this.listeners[ev].push(fn);
    }
    removeListener(ev, fn) {
        const list = this.listeners[ev], i = list.indexOf(fn);
        -1 !== i && (0 === i && 1 === list.length ? list.length = 0 : list.splice(i, 1));
    }
    emit(ev, code, signal) {
        if (this.emitted[ev]) return !1;
        this.emitted[ev] = !0;
        let ret = !1;
        for (const fn of this.listeners[ev]) ret = !0 === fn(code, signal) || ret;
        return "exit" === ev && (ret = this.emit("afterExit", code, signal) || ret), ret;
    }
}

class SignalExitBase {}

const process$1 = globalThis.process, {onExit: onExit, load: load, unload: unload} = (handler = processOk(process$1) ? new class extends SignalExitBase {
    #hupSig="win32" === process$1.platform ? "SIGINT" : "SIGHUP";
    #emitter=new Emitter;
    #process;
    #originalProcessEmit;
    #originalProcessReallyExit;
    #sigListeners={};
    #loaded=!1;
    constructor(process) {
        super(), this.#process = process, this.#sigListeners = {};
        for (const sig of signals) this.#sigListeners[sig] = () => {
            const listeners = this.#process.listeners(sig);
            let {count: count} = this.#emitter;
            const p = process;
            if ("object" == typeof p.__signal_exit_emitter__ && "number" == typeof p.__signal_exit_emitter__.count && (count += p.__signal_exit_emitter__.count), 
            listeners.length === count) {
                this.unload();
                const ret = this.#emitter.emit("exit", null, sig), s = "SIGHUP" === sig ? this.#hupSig : sig;
                ret || process.kill(process.pid, s);
            }
        };
        this.#originalProcessReallyExit = process.reallyExit, this.#originalProcessEmit = process.emit;
    }
    onExit(cb, opts) {
        if (!processOk(this.#process)) return () => {};
        !1 === this.#loaded && this.load();
        const ev = opts?.alwaysLast ? "afterExit" : "exit";
        return this.#emitter.on(ev, cb), () => {
            this.#emitter.removeListener(ev, cb), 0 === this.#emitter.listeners.exit.length && 0 === this.#emitter.listeners.afterExit.length && this.unload();
        };
    }
    load() {
        if (!this.#loaded) {
            this.#loaded = !0, this.#emitter.count += 1;
            for (const sig of signals) try {
                const fn = this.#sigListeners[sig];
                fn && this.#process.on(sig, fn);
            } catch (_) {}
            this.#process.emit = (ev, ...a) => this.#processEmit(ev, ...a), this.#process.reallyExit = code => this.#processReallyExit(code);
        }
    }
    unload() {
        this.#loaded && (this.#loaded = !1, signals.forEach((sig => {
            const listener = this.#sigListeners[sig];
            if (!listener) throw new Error("Listener not defined for signal: " + sig);
            try {
                this.#process.removeListener(sig, listener);
            } catch (_) {}
        })), this.#process.emit = this.#originalProcessEmit, this.#process.reallyExit = this.#originalProcessReallyExit, 
        this.#emitter.count -= 1);
    }
    #processReallyExit(code) {
        return processOk(this.#process) ? (this.#process.exitCode = code || 0, this.#emitter.emit("exit", this.#process.exitCode, null), 
        this.#originalProcessReallyExit.call(this.#process, this.#process.exitCode)) : 0;
    }
    #processEmit(ev, ...args) {
        const og = this.#originalProcessEmit;
        if ("exit" === ev && processOk(this.#process)) {
            "number" == typeof args[0] && (this.#process.exitCode = args[0]);
            const ret = og.call(this.#process, ev, ...args);
            return this.#emitter.emit("exit", this.#process.exitCode, null), ret;
        }
        return og.call(this.#process, ev, ...args);
    }
}(process$1) : new class extends SignalExitBase {
    onExit() {
        return () => {};
    }
    load() {}
    unload() {}
}, {
    onExit: (cb, opts) => handler.onExit(cb, opts),
    load: () => handler.load(),
    unload: () => handler.unload()
});

var handler;

const terminal = process$2.stderr.isTTY ? process$2.stderr : process$2.stdout.isTTY ? process$2.stdout : void 0, restoreCursor = terminal ? onetime((() => {
    onExit((() => {
        terminal.write("[?25h");
    }), {
        alwaysLast: !0
    });
})) : () => {};

let isHidden = !1;

const cliCursor = {
    show: (writableStream = process$2.stderr) => {
        writableStream.isTTY && (isHidden = !1, writableStream.write("[?25h"));
    },
    hide: (writableStream = process$2.stderr) => {
        writableStream.isTTY && (restoreCursor(), isHidden = !0, writableStream.write("[?25l"));
    },
    toggle: (force, writableStream) => {
        void 0 !== force && (isHidden = force), isHidden ? cliCursor.show(writableStream) : cliCursor.hide(writableStream);
    }
};

var cliSpinners$1, hasRequiredCliSpinners, require$$0 = {
    dots: {
        interval: 80,
        frames: [ "т ", "т ", "т Й", "т И", "т М", "т Д", "т І", "т Ї", "т ", "т " ]
    },
    dots2: {
        interval: 80,
        frames: [ "тЃО", "тЃН", "тЃЛ", "тЂП", "тЁП", "тЃ", "тЃЏ", "тЃЗ" ]
    },
    dots3: {
        interval: 80,
        frames: [ "т ", "т ", "т ", "т ", "т ", "т І", "т Д", "т В", "т Г", "т " ]
    },
    dots4: {
        interval: 80,
        frames: [ "т ", "т ", "т ", "т ", "т ", "т И", "т А", "т  ", "т А", "т И", "т ", "т ", "т ", "т " ]
    },
    dots5: {
        interval: 80,
        frames: [ "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т В", "т Д", "т І", "т ", "т ", "т ", "т ", "т ", "т ", "т " ]
    },
    dots6: {
        interval: 80,
        frames: [ "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т В", "т Д", "т Є", "т ", "т ", "т Є", "т Д", "т В", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т " ]
    },
    dots7: {
        interval: 80,
        frames: [ "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т І", "т Є", "т  ", "т  ", "т Є", "т І", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т " ]
    },
    dots8: {
        interval: 80,
        frames: [ "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т В", "т Д", "т Є", "т ", "т ", "т Є", "т  ", "т  ", "т Є", "т І", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т " ]
    },
    dots9: {
        interval: 80,
        frames: [ "тЂЙ", "тЂК", "тЂМ", "тЃИ", "тЃ", "тЁЇ", "тЁ", "тЁ" ]
    },
    dots10: {
        interval: 80,
        frames: [ "тЂ", "тЂ", "тЂ", "тЁ", "тЁ", "тЁ", "тЁ " ]
    },
    dots11: {
        interval: 100,
        frames: [ "т ", "т ", "т ", "тЁ", "тЂ", "т  ", "т ", "т " ]
    },
    dots12: {
        interval: 80,
        frames: [ "тЂт ", "тЁт ", "т т ", "тЂт ", "тЁт ", "т т ", "тЂт ", "тЁт ", "т т ", "тЂт ", "тЁт ", "т т ", "тЂт ", "тЁт ", "т т ", "т т ", "т т ", "т т ", "т т ", "т т Љ", "т тЂ", "т тЁ", "тЂт Љ", "тЁтЂ", "т тЁ", "тЂт Љ", "тЁтЂ", "т тЁ", "тЂт Ј", "тЁтЂ", "т тЁ", "тЂт  ", "тЁтЂ", "т тЁ", "тЂт ", "тЁт ", "т т ", "т т ", "т т ", "т т ", "т т ", "т т Љ", "т тЂ", "т тЁ", "т т Љ", "т тЂ", "т тЁ", "т т Љ", "т тЂ", "т тЁ", "т т Ј", "т тЂ", "т тЁ", "т т  ", "т тЂ", "т тЁ" ]
    },
    dots13: {
        interval: 80,
        frames: [ "тЃМ", "тЃЙ", "тЂЛ", "т П", "тЁ", "тЃ", "тЃЇ", "тЃЖ" ]
    },
    dots8Bit: {
        interval: 80,
        frames: [ "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "т ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "тЁ", "т  ", "т Ё", "т Ђ", "т Ѓ", "т Є", "т Ѕ", "т І", "т Ї", "тЁ ", "тЁЁ", "тЁЂ", "тЁЃ", "тЁЄ", "тЁЅ", "тЁІ", "тЁЇ", "т Ј", "т Љ", "т Њ", "т Ћ", "т Ќ", "т ­", "т Ў", "т Џ", "тЁЈ", "тЁЉ", "тЁЊ", "тЁЋ", "тЁЌ", "тЁ­", "тЁЎ", "тЁЏ", "т А", "т Б", "т В", "т Г", "т Д", "т Е", "т Ж", "т З", "тЁА", "тЁБ", "тЁВ", "тЁГ", "тЁД", "тЁЕ", "тЁЖ", "тЁЗ", "т И", "т Й", "т К", "т Л", "т М", "т Н", "т О", "т П", "тЁИ", "тЁЙ", "тЁК", "тЁЛ", "тЁМ", "тЁН", "тЁО", "тЁП", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЂ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЂ ", "тЂЁ", "тЂЂ", "тЂЃ", "тЂЄ", "тЂЅ", "тЂІ", "тЂЇ", "тЃ ", "тЃЁ", "тЃЂ", "тЃЃ", "тЃЄ", "тЃЅ", "тЃІ", "тЃЇ", "тЂЈ", "тЂЉ", "тЂЊ", "тЂЋ", "тЂЌ", "тЂ­", "тЂЎ", "тЂЏ", "тЃЈ", "тЃЉ", "тЃЊ", "тЃЋ", "тЃЌ", "тЃ­", "тЃЎ", "тЃЏ", "тЂА", "тЂБ", "тЂВ", "тЂГ", "тЂД", "тЂЕ", "тЂЖ", "тЂЗ", "тЃА", "тЃБ", "тЃВ", "тЃГ", "тЃД", "тЃЕ", "тЃЖ", "тЃЗ", "тЂИ", "тЂЙ", "тЂК", "тЂЛ", "тЂМ", "тЂН", "тЂО", "тЂП", "тЃИ", "тЃЙ", "тЃК", "тЃЛ", "тЃМ", "тЃН", "тЃО", "тЃП" ]
    },
    sand: {
        interval: 80,
        frames: [ "т ", "т ", "т ", "тЁ", "тЁ", "тЁ", "тЁ ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃ", "тЃЄ", "тЃЅ", "тЃІ", "тЃЎ", "тЃЖ", "тЃЗ", "тЃП", "тЁП", "т П", "тЂ", "т ", "тЁ", "т ", "т Ћ", "тЂ", "т ", "т ", "тЁ", "т ", "т ", "т Ё", "тЂ" ]
    },
    line: {
        interval: 130,
        frames: [ "-", "\\", "|", "/" ]
    },
    line2: {
        interval: 100,
        frames: [ "т ", "-", "т", "т", "т", "-" ]
    },
    pipe: {
        interval: 100,
        frames: [ "тЄ", "т", "тД", "т", "т", "т", "тЌ", "т" ]
    },
    simpleDots: {
        interval: 400,
        frames: [ ".  ", ".. ", "...", "   " ]
    },
    simpleDotsScrolling: {
        interval: 200,
        frames: [ ".  ", ".. ", "...", " ..", "  .", "   " ]
    },
    star: {
        interval: 70,
        frames: [ "тЖ", "тИ", "тЙ", "тК", "тЙ", "тЗ" ]
    },
    star2: {
        interval: 80,
        frames: [ "+", "x", "*" ]
    },
    flip: {
        interval: 70,
        frames: [ "_", "_", "_", "-", "`", "`", "'", "ТД", "-", "_", "_", "_" ]
    },
    hamburger: {
        interval: 100,
        frames: [ "тБ", "тВ", "тД" ]
    },
    growVertical: {
        interval: 120,
        frames: [ "т", "т", "т", "т", "т", "т", "т", "т", "т", "т" ]
    },
    growHorizontal: {
        interval: 120,
        frames: [ "т", "т", "т", "т", "т", "т", "т", "т", "т", "т", "т", "т" ]
    },
    balloon: {
        interval: 140,
        frames: [ " ", ".", "o", "O", "@", "*", " " ]
    },
    balloon2: {
        interval: 120,
        frames: [ ".", "o", "O", "ТА", "O", "o", "." ]
    },
    noise: {
        interval: 100,
        frames: [ "т", "т", "т" ]
    },
    bounce: {
        interval: 120,
        frames: [ "т ", "т ", "т ", "т " ]
    },
    boxBounce: {
        interval: 120,
        frames: [ "т", "т", "т", "т" ]
    },
    boxBounce2: {
        interval: 100,
        frames: [ "т", "т", "т", "т" ]
    },
    triangle: {
        interval: 50,
        frames: [ "тЂ", "тЃ", "тЄ", "тЅ" ]
    },
    binary: {
        interval: 80,
        frames: [ "010010", "001100", "100101", "111010", "111101", "010111", "101011", "111000", "110011", "110101" ]
    },
    arc: {
        interval: 100,
        frames: [ "т", "т ", "т", "т", "тЁ", "т" ]
    },
    circle: {
        interval: 120,
        frames: [ "тЁ", "т", "т " ]
    },
    squareCorners: {
        interval: 180,
        frames: [ "тА", "тГ", "тВ", "тБ" ]
    },
    circleQuarters: {
        interval: 120,
        frames: [ "тД", "тЗ", "тЖ", "тЕ" ]
    },
    circleHalves: {
        interval: 50,
        frames: [ "т", "т", "т", "т" ]
    },
    squish: {
        interval: 100,
        frames: [ "тЋ", "тЊ" ]
    },
    toggle: {
        interval: 250,
        frames: [ "тЖ", "тЗ" ]
    },
    toggle2: {
        interval: 80,
        frames: [ "тЋ", "тЊ" ]
    },
    toggle3: {
        interval: 120,
        frames: [ "тЁ", "т " ]
    },
    toggle4: {
        interval: 100,
        frames: [ "т ", "тЁ", "тЊ", "тЋ" ]
    },
    toggle5: {
        interval: 100,
        frames: [ "тЎ", "тЏ" ]
    },
    toggle6: {
        interval: 300,
        frames: [ "с", "с" ]
    },
    toggle7: {
        interval: 80,
        frames: [ "тІО", "тІП" ]
    },
    toggle8: {
        interval: 100,
        frames: [ "т", "т" ]
    },
    toggle9: {
        interval: 100,
        frames: [ "т", "т" ]
    },
    toggle10: {
        interval: 100,
        frames: [ "у", "у", "у" ]
    },
    toggle11: {
        interval: 50,
        frames: [ "тЇ", "тЇ" ]
    },
    toggle12: {
        interval: 120,
        frames: [ "т", "т" ]
    },
    toggle13: {
        interval: 80,
        frames: [ "=", "*", "-" ]
    },
    arrow: {
        interval: 100,
        frames: [ "т", "т", "т", "т", "т", "т", "т", "т" ]
    },
    arrow2: {
        interval: 80,
        frames: [ "тЌяИ ", "тяИ ", "тЁяИ ", "тяИ ", "тЌяИ ", "тяИ ", "тЌяИ ", "тяИ " ]
    },
    arrow3: {
        interval: 120,
        frames: [ "тЙтЙтЙтЙтЙ", "тИтЙтЙтЙтЙ", "тЙтИтЙтЙтЙ", "тЙтЙтИтЙтЙ", "тЙтЙтЙтИтЙ", "тЙтЙтЙтЙтИ" ]
    },
    bouncingBar: {
        interval: 80,
        frames: [ "[    ]", "[=   ]", "[==  ]", "[=== ]", "[====]", "[ ===]", "[  ==]", "[   =]", "[    ]", "[   =]", "[  ==]", "[ ===]", "[====]", "[=== ]", "[==  ]", "[=   ]" ]
    },
    bouncingBall: {
        interval: 80,
        frames: [ "( т    )", "(  т   )", "(   т  )", "(    т )", "(     т)", "(    т )", "(   т  )", "(  т   )", "( т    )", "(т     )" ]
    },
    smiley: {
        interval: 200,
        frames: [ "№ ", "№ " ]
    },
    monkey: {
        interval: 300,
        frames: [ "№ ", "№ ", "№ ", "№ " ]
    },
    hearts: {
        interval: 100,
        frames: [ "№ ", "№ ", "№ ", "№ ", "тЄяИ " ]
    },
    clock: {
        interval: 100,
        frames: [ "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ " ]
    },
    earth: {
        interval: 180,
        frames: [ "№ ", "№ ", "№ " ]
    },
    material: {
        interval: 17,
        frames: [ "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт", "тттттттттттттттттттт" ]
    },
    moon: {
        interval: 80,
        frames: [ "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ " ]
    },
    runner: {
        interval: 140,
        frames: [ "№Ж ", "№ " ]
    },
    pong: {
        interval: 80,
        frames: [ "тт        т", "тт        т", "т т       т", "т т        т", "т  тЁ     т", "т  т       т", "т   т     т", "т   т     т", "т    т    т", "т    т     т", "т     тЁ  т", "т     т    т", "т      т  т", "т      т  т", "т       т т", "т       т  т", "т       тЁт", "т      т   т", "т      т  т", "т     т   т", "т     т   т", "т    т     т", "т    тЁ   т", "т   т      т", "т   т     т", "т  т      т", "т  т      т", "т т        т", "т тЁ      т", "тт         т" ]
    },
    shark: {
        interval: 120,
        frames: [ "т|\\____________т", "т_|\\___________т", "т__|\\__________т", "т___|\\_________т", "т____|\\________т", "т_____|\\_______т", "т______|\\______т", "т_______|\\_____т", "т________|\\____т", "т_________|\\___т", "т__________|\\__т", "т___________|\\_т", "т____________|\\т", "т____________/|т", "т___________/|_т", "т__________/|__т", "т_________/|___т", "т________/|____т", "т_______/|_____т", "т______/|______т", "т_____/|_______т", "т____/|________т", "т___/|_________т", "т__/|__________т", "т_/|___________т", "т/|____________т" ]
    },
    dqpb: {
        interval: 100,
        frames: [ "d", "q", "p", "b" ]
    },
    weather: {
        interval: 100,
        frames: [ "тяИ ", "тяИ ", "тяИ ", "№Є ", "тяИ ", "№Ѕ ", "тяИ ", "№Ї ", "№Ј ", "№Ї ", "№Ј ", "№Ї ", "№Ј ", "т ", "№Ј ", "№Ї ", "№Ј ", "тяИ ", "№Ѕ ", "тяИ ", "№Є ", "тяИ ", "тяИ " ]
    },
    christmas: {
        interval: 400,
        frames: [ "№В", "№" ]
    },
    grenade: {
        interval: 80,
        frames: [ "и  ", "тВ  ", " ТД ", " тО ", "  тИ", "  тИ", "  |", "  т", "  т", " рЗД ", "  т", "   ", "   ", "   " ]
    },
    point: {
        interval: 125,
        frames: [ "ттт", "ттт", "ттт", "ттт", "ттт" ]
    },
    layer: {
        interval: 150,
        frames: [ "-", "=", "тЁ" ]
    },
    betaWave: {
        interval: 80,
        frames: [ "ЯЮВЮВЮВЮВЮВЮВ", "ЮВЯЮВЮВЮВЮВЮВ", "ЮВЮВЯЮВЮВЮВЮВ", "ЮВЮВЮВЯЮВЮВЮВ", "ЮВЮВЮВЮВЯЮВЮВ", "ЮВЮВЮВЮВЮВЯЮВ", "ЮВЮВЮВЮВЮВЮВЯ" ]
    },
    fingerDance: {
        interval: 160,
        frames: [ "№Є ", "№Є ", "№ ", "т ", "№Є ", "№ " ]
    },
    fistBump: {
        interval: 80,
        frames: [ "№Єуууу№Є ", "№Єуууу№Є ", "№Єуууу№Є ", "у№Єуу№Єу ", "уу№Є№Єуу ", "у№ЄтЈ№Єуу ", "№ЄутЈу№Єу " ]
    },
    soccerHeader: {
        interval: 80,
        frames: [ " №ЇтНяИ       №Ї ", "№Ї  тНяИ      №Ї ", "№Ї   тНяИ     №Ї ", "№Ї    тНяИ    №Ї ", "№Ї     тНяИ   №Ї ", "№Ї      тНяИ  №Ї ", "№Ї       тНяИ№Ї  ", "№Ї      тНяИ  №Ї ", "№Ї     тНяИ   №Ї ", "№Ї    тНяИ    №Ї ", "№Ї   тНяИ     №Ї ", "№Ї  тНяИ      №Ї " ]
    },
    mindblown: {
        interval: 160,
        frames: [ "№ ", "№ ", "№Ў ", "№Ў ", "№І ", "№І ", "№Ї ", "№Ї ", "№ЄЏ ", "№Ѕ ", "тЈ ", "у ", "у ", "у " ]
    },
    speaker: {
        interval: 160,
        frames: [ "№ ", "№ ", "№ ", "№ " ]
    },
    orangePulse: {
        interval: 100,
        frames: [ "№И ", "№Ж ", "№  ", "№  ", "№Ж " ]
    },
    bluePulse: {
        interval: 100,
        frames: [ "№Й ", "№З ", "№Е ", "№Е ", "№З " ]
    },
    orangeBluePulse: {
        interval: 100,
        frames: [ "№И ", "№Ж ", "№  ", "№  ", "№Ж ", "№Й ", "№З ", "№Е ", "№Е ", "№З " ]
    },
    timeTravel: {
        interval: 100,
        frames: [ "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ ", "№ " ]
    },
    aesthetic: {
        interval: 80,
        frames: [ "тАтБтБтБтБтБтБ", "тАтАтБтБтБтБтБ", "тАтАтАтБтБтБтБ", "тАтАтАтАтБтБтБ", "тАтАтАтАтАтБтБ", "тАтАтАтАтАтАтБ", "тАтАтАтАтАтАтА", "тАтБтБтБтБтБтБ" ]
    },
    dwarfFortress: {
        interval: 80,
        frames: [ " ттттттТЃТЃТЃ  ", "тКттттттТЃТЃТЃ  ", "тКттттттТЃТЃТЃ  ", "тКттттттТЃТЃТЃ  ", "тКттттттТЃТЃТЃ  ", "тКттттттТЃТЃТЃ  ", "тКттттттТЃТЃТЃ  ", "тКттттттТЃТЃТЃ  ", "тКттттттТЃТЃТЃ  ", "тК тттттТЃТЃТЃ  ", " тКтттттТЃТЃТЃ  ", " тКтттттТЃТЃТЃ  ", " тКтттттТЃТЃТЃ  ", " тКтттттТЃТЃТЃ  ", " тКтттттТЃТЃТЃ  ", " тКтттттТЃТЃТЃ  ", " тКтттттТЃТЃТЃ  ", " тКтттттТЃТЃТЃ  ", " тК ттттТЃТЃТЃ  ", "  тКттттТЃТЃТЃ  ", "  тКттттТЃТЃТЃ  ", "  тКттттТЃТЃТЃ  ", "  тКттттТЃТЃТЃ  ", "  тКттттТЃТЃТЃ  ", "  тКттттТЃТЃТЃ  ", "  тКттттТЃТЃТЃ  ", "  тКттттТЃТЃТЃ  ", "  тК тттТЃТЃТЃ  ", "   тКтттТЃТЃТЃ  ", "   тКтттТЃТЃТЃ  ", "   тКтттТЃТЃТЃ  ", "   тКтттТЃТЃТЃ  ", "   тКтттТЃТЃТЃ  ", "   тКтттТЃТЃТЃ  ", "   тКтттТЃТЃТЃ  ", "   тКтттТЃТЃТЃ  ", "   тК ттТЃТЃТЃ  ", "    тКттТЃТЃТЃ  ", "    тКттТЃТЃТЃ  ", "    тКттТЃТЃТЃ  ", "    тКттТЃТЃТЃ  ", "    тКттТЃТЃТЃ  ", "    тКттТЃТЃТЃ  ", "    тКттТЃТЃТЃ  ", "    тКттТЃТЃТЃ  ", "    тК тТЃТЃТЃ  ", "     тКтТЃТЃТЃ  ", "     тКтТЃТЃТЃ  ", "     тКтТЃТЃТЃ  ", "     тКтТЃТЃТЃ  ", "     тКтТЃТЃТЃ  ", "     тКтТЃТЃТЃ  ", "     тКтТЃТЃТЃ  ", "     тКтТЃТЃТЃ  ", "     тК ТЃТЃТЃ  ", "      тКТЃТЃТЃ  ", "      тКТЃТЃТЃ  ", "      тКтТЃТЃ  ", "      тКтТЃТЃ  ", "      тКтТЃТЃ  ", "      тКтТЃТЃ  ", "      тКтТЃТЃ  ", "      тКтТЃТЃ  ", "      тК ТЃТЃ  ", "       тКТЃТЃ  ", "       тКТЃТЃ  ", "       тКтТЃ  ", "       тКтТЃ  ", "       тКтТЃ  ", "       тКтТЃ  ", "       тКтТЃ  ", "       тКтТЃ  ", "       тК ТЃ  ", "        тКТЃ  ", "        тКТЃ  ", "        тКт  ", "        тКт  ", "        тКт  ", "        тКт  ", "        тКт  ", "        тКт  ", "        тК   ", "        тК  &", "        тК тМ&", "       тК тМ &", "       тКтМ  &", "      тКтМ  & ", "      тМ   & ", "     тК   &  ", "    тМ    &  ", "   тК    &   ", "  тМ     &   ", " тК     &    ", "тМ      &    ", "      &     ", "      &     ", "     &   т  ", "     &   т  ", "    &    т  ", "    &    ТЃ  ", "   &    тТЃ  ", "   &    тТЃ  ", "  &     тТЃ  ", "  &     ТЃТЃ  ", " &     тТЃТЃ  ", " &     тТЃТЃ  ", "&      тТЃТЃ  ", "&      ТЃТЃТЃ  ", "      тТЃТЃТЃ  ", "      тТЃТЃТЃ  ", "      тТЃТЃТЃ  ", "      тТЃТЃТЃ  ", "     ттТЃТЃТЃ  ", "     ттТЃТЃТЃ  ", "     ттТЃТЃТЃ  ", "     ттТЃТЃТЃ  ", "    тттТЃТЃТЃ  ", "    тттТЃТЃТЃ  ", "    тттТЃТЃТЃ  ", "    тттТЃТЃТЃ  ", "   ттттТЃТЃТЃ  ", "   ттттТЃТЃТЃ  ", "   ттттТЃТЃТЃ  ", "   ттттТЃТЃТЃ  ", "  тттттТЃТЃТЃ  ", "  тттттТЃТЃТЃ  ", "  тттттТЃТЃТЃ  ", "  тттттТЃТЃТЃ  ", " ттттттТЃТЃТЃ  ", " ттттттТЃТЃТЃ  ", " ттттттТЃТЃТЃ  ", " ттттттТЃТЃТЃ  ", " ттттттТЃТЃТЃ  " ]
    }
};

var cliSpinners = getDefaultExportFromCjs(function() {
    if (hasRequiredCliSpinners) return cliSpinners$1;
    hasRequiredCliSpinners = 1;
    const spinners = Object.assign({}, require$$0), spinnersList = Object.keys(spinners);
    return Object.defineProperty(spinners, "random", {
        get() {
            const randomIndex = Math.floor(Math.random() * spinnersList.length), spinnerName = spinnersList[randomIndex];
            return spinners[spinnerName];
        }
    }), cliSpinners$1 = spinners;
}());

const main$1 = {
    info: chalk.blue("тЙ"),
    success: chalk.green("т"),
    warning: chalk.yellow("т "),
    error: chalk.red("т")
}, fallback = {
    info: chalk.blue("i"),
    success: chalk.green("т"),
    warning: chalk.yellow("тМ"),
    error: chalk.red("У")
}, logSymbols = ("win32" !== process$2.platform ? "linux" !== process$2.env.TERM : Boolean(process$2.env.CI) || Boolean(process$2.env.WT_SESSION) || Boolean(process$2.env.TERMINUS_SUBLIME) || "{cmd::Cmder}" === process$2.env.ConEmuTask || "Terminus-Sublime" === process$2.env.TERM_PROGRAM || "vscode" === process$2.env.TERM_PROGRAM || "xterm-256color" === process$2.env.TERM || "alacritty" === process$2.env.TERM || "JetBrains-JediTerm" === process$2.env.TERMINAL_EMULATOR) ? main$1 : fallback;

const regex = function({onlyFirst: onlyFirst = !1} = {}) {
    const pattern = [ "[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]+)*|[a-zA-Z\\d]+(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?(?:\\u0007|\\u001B\\u005C|\\u009C))", "(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PR-TZcf-nq-uy=><~]))" ].join("|");
    return new RegExp(pattern, onlyFirst ? void 0 : "g");
}();

function stripAnsi(string) {
    if ("string" != typeof string) throw new TypeError(`Expected a \`string\`, got \`${typeof string}\``);
    return string.replace(regex, "");
}

function eastAsianWidth(codePoint, {ambiguousAsWide: ambiguousAsWide = !1} = {}) {
    return function(codePoint) {
        if (!Number.isSafeInteger(codePoint)) throw new TypeError(`Expected a code point, got \`${typeof codePoint}\`.`);
    }(codePoint), 12288 === (x = codePoint) || x >= 65281 && x <= 65376 || x >= 65504 && x <= 65510 || function(x) {
        return x >= 4352 && x <= 4447 || 8986 === x || 8987 === x || 9001 === x || 9002 === x || x >= 9193 && x <= 9196 || 9200 === x || 9203 === x || 9725 === x || 9726 === x || 9748 === x || 9749 === x || x >= 9800 && x <= 9811 || 9855 === x || 9875 === x || 9889 === x || 9898 === x || 9899 === x || 9917 === x || 9918 === x || 9924 === x || 9925 === x || 9934 === x || 9940 === x || 9962 === x || 9970 === x || 9971 === x || 9973 === x || 9978 === x || 9981 === x || 9989 === x || 9994 === x || 9995 === x || 10024 === x || 10060 === x || 10062 === x || x >= 10067 && x <= 10069 || 10071 === x || x >= 10133 && x <= 10135 || 10160 === x || 10175 === x || 11035 === x || 11036 === x || 11088 === x || 11093 === x || x >= 11904 && x <= 11929 || x >= 11931 && x <= 12019 || x >= 12032 && x <= 12245 || x >= 12272 && x <= 12287 || x >= 12289 && x <= 12350 || x >= 12353 && x <= 12438 || x >= 12441 && x <= 12543 || x >= 12549 && x <= 12591 || x >= 12593 && x <= 12686 || x >= 12688 && x <= 12771 || x >= 12783 && x <= 12830 || x >= 12832 && x <= 12871 || x >= 12880 && x <= 19903 || x >= 19968 && x <= 42124 || x >= 42128 && x <= 42182 || x >= 43360 && x <= 43388 || x >= 44032 && x <= 55203 || x >= 63744 && x <= 64255 || x >= 65040 && x <= 65049 || x >= 65072 && x <= 65106 || x >= 65108 && x <= 65126 || x >= 65128 && x <= 65131 || x >= 94176 && x <= 94180 || 94192 === x || 94193 === x || x >= 94208 && x <= 100343 || x >= 100352 && x <= 101589 || x >= 101632 && x <= 101640 || x >= 110576 && x <= 110579 || x >= 110581 && x <= 110587 || 110589 === x || 110590 === x || x >= 110592 && x <= 110882 || 110898 === x || x >= 110928 && x <= 110930 || 110933 === x || x >= 110948 && x <= 110951 || x >= 110960 && x <= 111355 || 126980 === x || 127183 === x || 127374 === x || x >= 127377 && x <= 127386 || x >= 127488 && x <= 127490 || x >= 127504 && x <= 127547 || x >= 127552 && x <= 127560 || 127568 === x || 127569 === x || x >= 127584 && x <= 127589 || x >= 127744 && x <= 127776 || x >= 127789 && x <= 127797 || x >= 127799 && x <= 127868 || x >= 127870 && x <= 127891 || x >= 127904 && x <= 127946 || x >= 127951 && x <= 127955 || x >= 127968 && x <= 127984 || 127988 === x || x >= 127992 && x <= 128062 || 128064 === x || x >= 128066 && x <= 128252 || x >= 128255 && x <= 128317 || x >= 128331 && x <= 128334 || x >= 128336 && x <= 128359 || 128378 === x || 128405 === x || 128406 === x || 128420 === x || x >= 128507 && x <= 128591 || x >= 128640 && x <= 128709 || 128716 === x || x >= 128720 && x <= 128722 || x >= 128725 && x <= 128727 || x >= 128732 && x <= 128735 || 128747 === x || 128748 === x || x >= 128756 && x <= 128764 || x >= 128992 && x <= 129003 || 129008 === x || x >= 129292 && x <= 129338 || x >= 129340 && x <= 129349 || x >= 129351 && x <= 129535 || x >= 129648 && x <= 129660 || x >= 129664 && x <= 129672 || x >= 129680 && x <= 129725 || x >= 129727 && x <= 129733 || x >= 129742 && x <= 129755 || x >= 129760 && x <= 129768 || x >= 129776 && x <= 129784 || x >= 131072 && x <= 196605 || x >= 196608 && x <= 262141;
    }(codePoint) || ambiguousAsWide && function(x) {
        return 161 === x || 164 === x || 167 === x || 168 === x || 170 === x || 173 === x || 174 === x || x >= 176 && x <= 180 || x >= 182 && x <= 186 || x >= 188 && x <= 191 || 198 === x || 208 === x || 215 === x || 216 === x || x >= 222 && x <= 225 || 230 === x || x >= 232 && x <= 234 || 236 === x || 237 === x || 240 === x || 242 === x || 243 === x || x >= 247 && x <= 250 || 252 === x || 254 === x || 257 === x || 273 === x || 275 === x || 283 === x || 294 === x || 295 === x || 299 === x || x >= 305 && x <= 307 || 312 === x || x >= 319 && x <= 322 || 324 === x || x >= 328 && x <= 331 || 333 === x || 338 === x || 339 === x || 358 === x || 359 === x || 363 === x || 462 === x || 464 === x || 466 === x || 468 === x || 470 === x || 472 === x || 474 === x || 476 === x || 593 === x || 609 === x || 708 === x || 711 === x || x >= 713 && x <= 715 || 717 === x || 720 === x || x >= 728 && x <= 731 || 733 === x || 735 === x || x >= 768 && x <= 879 || x >= 913 && x <= 929 || x >= 931 && x <= 937 || x >= 945 && x <= 961 || x >= 963 && x <= 969 || 1025 === x || x >= 1040 && x <= 1103 || 1105 === x || 8208 === x || x >= 8211 && x <= 8214 || 8216 === x || 8217 === x || 8220 === x || 8221 === x || x >= 8224 && x <= 8226 || x >= 8228 && x <= 8231 || 8240 === x || 8242 === x || 8243 === x || 8245 === x || 8251 === x || 8254 === x || 8308 === x || 8319 === x || x >= 8321 && x <= 8324 || 8364 === x || 8451 === x || 8453 === x || 8457 === x || 8467 === x || 8470 === x || 8481 === x || 8482 === x || 8486 === x || 8491 === x || 8531 === x || 8532 === x || x >= 8539 && x <= 8542 || x >= 8544 && x <= 8555 || x >= 8560 && x <= 8569 || 8585 === x || x >= 8592 && x <= 8601 || 8632 === x || 8633 === x || 8658 === x || 8660 === x || 8679 === x || 8704 === x || 8706 === x || 8707 === x || 8711 === x || 8712 === x || 8715 === x || 8719 === x || 8721 === x || 8725 === x || 8730 === x || x >= 8733 && x <= 8736 || 8739 === x || 8741 === x || x >= 8743 && x <= 8748 || 8750 === x || x >= 8756 && x <= 8759 || 8764 === x || 8765 === x || 8776 === x || 8780 === x || 8786 === x || 8800 === x || 8801 === x || x >= 8804 && x <= 8807 || 8810 === x || 8811 === x || 8814 === x || 8815 === x || 8834 === x || 8835 === x || 8838 === x || 8839 === x || 8853 === x || 8857 === x || 8869 === x || 8895 === x || 8978 === x || x >= 9312 && x <= 9449 || x >= 9451 && x <= 9547 || x >= 9552 && x <= 9587 || x >= 9600 && x <= 9615 || x >= 9618 && x <= 9621 || 9632 === x || 9633 === x || x >= 9635 && x <= 9641 || 9650 === x || 9651 === x || 9654 === x || 9655 === x || 9660 === x || 9661 === x || 9664 === x || 9665 === x || x >= 9670 && x <= 9672 || 9675 === x || x >= 9678 && x <= 9681 || x >= 9698 && x <= 9701 || 9711 === x || 9733 === x || 9734 === x || 9737 === x || 9742 === x || 9743 === x || 9756 === x || 9758 === x || 9792 === x || 9794 === x || 9824 === x || 9825 === x || x >= 9827 && x <= 9829 || x >= 9831 && x <= 9834 || 9836 === x || 9837 === x || 9839 === x || 9886 === x || 9887 === x || 9919 === x || x >= 9926 && x <= 9933 || x >= 9935 && x <= 9939 || x >= 9941 && x <= 9953 || 9955 === x || 9960 === x || 9961 === x || x >= 9963 && x <= 9969 || 9972 === x || x >= 9974 && x <= 9977 || 9979 === x || 9980 === x || 9982 === x || 9983 === x || 10045 === x || x >= 10102 && x <= 10111 || x >= 11094 && x <= 11097 || x >= 12872 && x <= 12879 || x >= 57344 && x <= 63743 || x >= 65024 && x <= 65039 || 65533 === x || x >= 127232 && x <= 127242 || x >= 127248 && x <= 127277 || x >= 127280 && x <= 127337 || x >= 127344 && x <= 127373 || 127375 === x || 127376 === x || x >= 127387 && x <= 127404 || x >= 917760 && x <= 917999 || x >= 983040 && x <= 1048573 || x >= 1048576 && x <= 1114109;
    }(codePoint) ? 2 : 1;
    var x;
}

const segmenter = new Intl.Segmenter, defaultIgnorableCodePointRegex = /^\p{Default_Ignorable_Code_Point}$/u;

function stringWidth(string, options = {}) {
    if ("string" != typeof string || 0 === string.length) return 0;
    const {ambiguousIsNarrow: ambiguousIsNarrow = !0, countAnsiEscapeCodes: countAnsiEscapeCodes = !1} = options;
    if (countAnsiEscapeCodes || (string = stripAnsi(string)), 0 === string.length) return 0;
    let width = 0;
    const eastAsianWidthOptions = {
        ambiguousAsWide: !ambiguousIsNarrow
    };
    for (const {segment: character} of segmenter.segment(string)) {
        const codePoint = character.codePointAt(0);
        codePoint <= 31 || codePoint >= 127 && codePoint <= 159 || (codePoint >= 8203 && codePoint <= 8207 || 65279 === codePoint || codePoint >= 768 && codePoint <= 879 || codePoint >= 6832 && codePoint <= 6911 || codePoint >= 7616 && codePoint <= 7679 || codePoint >= 8400 && codePoint <= 8447 || codePoint >= 65056 && codePoint <= 65071 || codePoint >= 55296 && codePoint <= 57343 || codePoint >= 65024 && codePoint <= 65039 || defaultIgnorableCodePointRegex.test(character) || (/[#*0-9]\uFE0F?\u20E3|[\xA9\xAE\u203C\u2049\u2122\u2139\u2194-\u2199\u21A9\u21AA\u231A\u231B\u2328\u23CF\u23ED-\u23EF\u23F1\u23F2\u23F8-\u23FA\u24C2\u25AA\u25AB\u25B6\u25C0\u25FB\u25FC\u25FE\u2600-\u2604\u260E\u2611\u2614\u2615\u2618\u2620\u2622\u2623\u2626\u262A\u262E\u262F\u2638-\u263A\u2640\u2642\u2648-\u2653\u265F\u2660\u2663\u2665\u2666\u2668\u267B\u267E\u267F\u2692\u2694-\u2697\u2699\u269B\u269C\u26A0\u26A7\u26AA\u26B0\u26B1\u26BD\u26BE\u26C4\u26C8\u26CF\u26D1\u26E9\u26F0-\u26F5\u26F7\u26F8\u26FA\u2702\u2708\u2709\u270F\u2712\u2714\u2716\u271D\u2721\u2733\u2734\u2744\u2747\u2757\u2763\u27A1\u2934\u2935\u2B05-\u2B07\u2B1B\u2B1C\u2B55\u3030\u303D\u3297\u3299]\uFE0F?|[\u261D\u270C\u270D](?:\uD83C[\uDFFB-\uDFFF]|\uFE0F)?|[\u270A\u270B](?:\uD83C[\uDFFB-\uDFFF])?|[\u23E9-\u23EC\u23F0\u23F3\u25FD\u2693\u26A1\u26AB\u26C5\u26CE\u26D4\u26EA\u26FD\u2705\u2728\u274C\u274E\u2753-\u2755\u2795-\u2797\u27B0\u27BF\u2B50]|\u26D3\uFE0F?(?:\u200D\uD83D\uDCA5)?|\u26F9(?:\uD83C[\uDFFB-\uDFFF]|\uFE0F)?(?:\u200D[\u2640\u2642]\uFE0F?)?|\u2764\uFE0F?(?:\u200D(?:\uD83D\uDD25|\uD83E\uDE79))?|\uD83C(?:[\uDC04\uDD70\uDD71\uDD7E\uDD7F\uDE02\uDE37\uDF21\uDF24-\uDF2C\uDF36\uDF7D\uDF96\uDF97\uDF99-\uDF9B\uDF9E\uDF9F\uDFCD\uDFCE\uDFD4-\uDFDF\uDFF5\uDFF7]\uFE0F?|[\uDF85\uDFC2\uDFC7](?:\uD83C[\uDFFB-\uDFFF])?|[\uDFC4\uDFCA](?:\uD83C[\uDFFB-\uDFFF])?(?:\u200D[\u2640\u2642]\uFE0F?)?|[\uDFCB\uDFCC](?:\uD83C[\uDFFB-\uDFFF]|\uFE0F)?(?:\u200D[\u2640\u2642]\uFE0F?)?|[\uDCCF\uDD8E\uDD91-\uDD9A\uDE01\uDE1A\uDE2F\uDE32-\uDE36\uDE38-\uDE3A\uDE50\uDE51\uDF00-\uDF20\uDF2D-\uDF35\uDF37-\uDF43\uDF45-\uDF4A\uDF4C-\uDF7C\uDF7E-\uDF84\uDF86-\uDF93\uDFA0-\uDFC1\uDFC5\uDFC6\uDFC8\uDFC9\uDFCF-\uDFD3\uDFE0-\uDFF0\uDFF8-\uDFFF]|\uDDE6\uD83C[\uDDE8-\uDDEC\uDDEE\uDDF1\uDDF2\uDDF4\uDDF6-\uDDFA\uDDFC\uDDFD\uDDFF]|\uDDE7\uD83C[\uDDE6\uDDE7\uDDE9-\uDDEF\uDDF1-\uDDF4\uDDF6-\uDDF9\uDDFB\uDDFC\uDDFE\uDDFF]|\uDDE8\uD83C[\uDDE6\uDDE8\uDDE9\uDDEB-\uDDEE\uDDF0-\uDDF7\uDDFA-\uDDFF]|\uDDE9\uD83C[\uDDEA\uDDEC\uDDEF\uDDF0\uDDF2\uDDF4\uDDFF]|\uDDEA\uD83C[\uDDE6\uDDE8\uDDEA\uDDEC\uDDED\uDDF7-\uDDFA]|\uDDEB\uD83C[\uDDEE-\uDDF0\uDDF2\uDDF4\uDDF7]|\uDDEC\uD83C[\uDDE6\uDDE7\uDDE9-\uDDEE\uDDF1-\uDDF3\uDDF5-\uDDFA\uDDFC\uDDFE]|\uDDED\uD83C[\uDDF0\uDDF2\uDDF3\uDDF7\uDDF9\uDDFA]|\uDDEE\uD83C[\uDDE8-\uDDEA\uDDF1-\uDDF4\uDDF6-\uDDF9]|\uDDEF\uD83C[\uDDEA\uDDF2\uDDF4\uDDF5]|\uDDF0\uD83C[\uDDEA\uDDEC-\uDDEE\uDDF2\uDDF3\uDDF5\uDDF7\uDDFC\uDDFE\uDDFF]|\uDDF1\uD83C[\uDDE6-\uDDE8\uDDEE\uDDF0\uDDF7-\uDDFB\uDDFE]|\uDDF2\uD83C[\uDDE6\uDDE8-\uDDED\uDDF0-\uDDFF]|\uDDF3\uD83C[\uDDE6\uDDE8\uDDEA-\uDDEC\uDDEE\uDDF1\uDDF4\uDDF5\uDDF7\uDDFA\uDDFF]|\uDDF4\uD83C\uDDF2|\uDDF5\uD83C[\uDDE6\uDDEA-\uDDED\uDDF0-\uDDF3\uDDF7-\uDDF9\uDDFC\uDDFE]|\uDDF6\uD83C\uDDE6|\uDDF7\uD83C[\uDDEA\uDDF4\uDDF8\uDDFA\uDDFC]|\uDDF8\uD83C[\uDDE6-\uDDEA\uDDEC-\uDDF4\uDDF7-\uDDF9\uDDFB\uDDFD-\uDDFF]|\uDDF9\uD83C[\uDDE6\uDDE8\uDDE9\uDDEB-\uDDED\uDDEF-\uDDF4\uDDF7\uDDF9\uDDFB\uDDFC\uDDFF]|\uDDFA\uD83C[\uDDE6\uDDEC\uDDF2\uDDF3\uDDF8\uDDFE\uDDFF]|\uDDFB\uD83C[\uDDE6\uDDE8\uDDEA\uDDEC\uDDEE\uDDF3\uDDFA]|\uDDFC\uD83C[\uDDEB\uDDF8]|\uDDFD\uD83C\uDDF0|\uDDFE\uD83C[\uDDEA\uDDF9]|\uDDFF\uD83C[\uDDE6\uDDF2\uDDFC]|\uDF44(?:\u200D\uD83D\uDFEB)?|\uDF4B(?:\u200D\uD83D\uDFE9)?|\uDFC3(?:\uD83C[\uDFFB-\uDFFF])?(?:\u200D(?:[\u2640\u2642]\uFE0F?(?:\u200D\u27A1\uFE0F?)?|\u27A1\uFE0F?))?|\uDFF3\uFE0F?(?:\u200D(?:\u26A7\uFE0F?|\uD83C\uDF08))?|\uDFF4(?:\u200D\u2620\uFE0F?|\uDB40\uDC67\uDB40\uDC62\uDB40(?:\uDC65\uDB40\uDC6E\uDB40\uDC67|\uDC73\uDB40\uDC63\uDB40\uDC74|\uDC77\uDB40\uDC6C\uDB40\uDC73)\uDB40\uDC7F)?)|\uD83D(?:[\uDC3F\uDCFD\uDD49\uDD4A\uDD6F\uDD70\uDD73\uDD76-\uDD79\uDD87\uDD8A-\uDD8D\uDDA5\uDDA8\uDDB1\uDDB2\uDDBC\uDDC2-\uDDC4\uDDD1-\uDDD3\uDDDC-\uDDDE\uDDE1\uDDE3\uDDE8\uDDEF\uDDF3\uDDFA\uDECB\uDECD-\uDECF\uDEE0-\uDEE5\uDEE9\uDEF0\uDEF3]\uFE0F?|[\uDC42\uDC43\uDC46-\uDC50\uDC66\uDC67\uDC6B-\uDC6D\uDC72\uDC74-\uDC76\uDC78\uDC7C\uDC83\uDC85\uDC8F\uDC91\uDCAA\uDD7A\uDD95\uDD96\uDE4C\uDE4F\uDEC0\uDECC](?:\uD83C[\uDFFB-\uDFFF])?|[\uDC6E\uDC70\uDC71\uDC73\uDC77\uDC81\uDC82\uDC86\uDC87\uDE45-\uDE47\uDE4B\uDE4D\uDE4E\uDEA3\uDEB4\uDEB5](?:\uD83C[\uDFFB-\uDFFF])?(?:\u200D[\u2640\u2642]\uFE0F?)?|[\uDD74\uDD90](?:\uD83C[\uDFFB-\uDFFF]|\uFE0F)?|[\uDC00-\uDC07\uDC09-\uDC14\uDC16-\uDC25\uDC27-\uDC3A\uDC3C-\uDC3E\uDC40\uDC44\uDC45\uDC51-\uDC65\uDC6A\uDC79-\uDC7B\uDC7D-\uDC80\uDC84\uDC88-\uDC8E\uDC90\uDC92-\uDCA9\uDCAB-\uDCFC\uDCFF-\uDD3D\uDD4B-\uDD4E\uDD50-\uDD67\uDDA4\uDDFB-\uDE2D\uDE2F-\uDE34\uDE37-\uDE41\uDE43\uDE44\uDE48-\uDE4A\uDE80-\uDEA2\uDEA4-\uDEB3\uDEB7-\uDEBF\uDEC1-\uDEC5\uDED0-\uDED2\uDED5-\uDED7\uDEDC-\uDEDF\uDEEB\uDEEC\uDEF4-\uDEFC\uDFE0-\uDFEB\uDFF0]|\uDC08(?:\u200D\u2B1B)?|\uDC15(?:\u200D\uD83E\uDDBA)?|\uDC26(?:\u200D(?:\u2B1B|\uD83D\uDD25))?|\uDC3B(?:\u200D\u2744\uFE0F?)?|\uDC41\uFE0F?(?:\u200D\uD83D\uDDE8\uFE0F?)?|\uDC68(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:\uDC8B\u200D\uD83D)?\uDC68|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D(?:[\uDC68\uDC69]\u200D\uD83D(?:\uDC66(?:\u200D\uD83D\uDC66)?|\uDC67(?:\u200D\uD83D[\uDC66\uDC67])?)|[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uDC66(?:\u200D\uD83D\uDC66)?|\uDC67(?:\u200D\uD83D[\uDC66\uDC67])?)|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]))|\uD83C(?:\uDFFB(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:\uDC8B\u200D\uD83D)?\uDC68\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D\uDC68\uD83C[\uDFFC-\uDFFF])))?|\uDFFC(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:\uDC8B\u200D\uD83D)?\uDC68\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D\uDC68\uD83C[\uDFFB\uDFFD-\uDFFF])))?|\uDFFD(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:\uDC8B\u200D\uD83D)?\uDC68\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D\uDC68\uD83C[\uDFFB\uDFFC\uDFFE\uDFFF])))?|\uDFFE(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:\uDC8B\u200D\uD83D)?\uDC68\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D\uDC68\uD83C[\uDFFB-\uDFFD\uDFFF])))?|\uDFFF(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:\uDC8B\u200D\uD83D)?\uDC68\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D\uDC68\uD83C[\uDFFB-\uDFFE])))?))?|\uDC69(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:\uDC8B\u200D\uD83D)?[\uDC68\uDC69]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D(?:[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uDC66(?:\u200D\uD83D\uDC66)?|\uDC67(?:\u200D\uD83D[\uDC66\uDC67])?|\uDC69\u200D\uD83D(?:\uDC66(?:\u200D\uD83D\uDC66)?|\uDC67(?:\u200D\uD83D[\uDC66\uDC67])?))|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]))|\uD83C(?:\uDFFB(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:[\uDC68\uDC69]|\uDC8B\u200D\uD83D[\uDC68\uDC69])\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D[\uDC68\uDC69]\uD83C[\uDFFC-\uDFFF])))?|\uDFFC(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:[\uDC68\uDC69]|\uDC8B\u200D\uD83D[\uDC68\uDC69])\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D[\uDC68\uDC69]\uD83C[\uDFFB\uDFFD-\uDFFF])))?|\uDFFD(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:[\uDC68\uDC69]|\uDC8B\u200D\uD83D[\uDC68\uDC69])\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D[\uDC68\uDC69]\uD83C[\uDFFB\uDFFC\uDFFE\uDFFF])))?|\uDFFE(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:[\uDC68\uDC69]|\uDC8B\u200D\uD83D[\uDC68\uDC69])\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D[\uDC68\uDC69]\uD83C[\uDFFB-\uDFFD\uDFFF])))?|\uDFFF(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D\uD83D(?:[\uDC68\uDC69]|\uDC8B\u200D\uD83D[\uDC68\uDC69])\uD83C[\uDFFB-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83D[\uDC68\uDC69]\uD83C[\uDFFB-\uDFFE])))?))?|\uDC6F(?:\u200D[\u2640\u2642]\uFE0F?)?|\uDD75(?:\uD83C[\uDFFB-\uDFFF]|\uFE0F)?(?:\u200D[\u2640\u2642]\uFE0F?)?|\uDE2E(?:\u200D\uD83D\uDCA8)?|\uDE35(?:\u200D\uD83D\uDCAB)?|\uDE36(?:\u200D\uD83C\uDF2B\uFE0F?)?|\uDE42(?:\u200D[\u2194\u2195]\uFE0F?)?|\uDEB6(?:\uD83C[\uDFFB-\uDFFF])?(?:\u200D(?:[\u2640\u2642]\uFE0F?(?:\u200D\u27A1\uFE0F?)?|\u27A1\uFE0F?))?)|\uD83E(?:[\uDD0C\uDD0F\uDD18-\uDD1F\uDD30-\uDD34\uDD36\uDD77\uDDB5\uDDB6\uDDBB\uDDD2\uDDD3\uDDD5\uDEC3-\uDEC5\uDEF0\uDEF2-\uDEF8](?:\uD83C[\uDFFB-\uDFFF])?|[\uDD26\uDD35\uDD37-\uDD39\uDD3D\uDD3E\uDDB8\uDDB9\uDDCD\uDDCF\uDDD4\uDDD6-\uDDDD](?:\uD83C[\uDFFB-\uDFFF])?(?:\u200D[\u2640\u2642]\uFE0F?)?|[\uDDDE\uDDDF](?:\u200D[\u2640\u2642]\uFE0F?)?|[\uDD0D\uDD0E\uDD10-\uDD17\uDD20-\uDD25\uDD27-\uDD2F\uDD3A\uDD3F-\uDD45\uDD47-\uDD76\uDD78-\uDDB4\uDDB7\uDDBA\uDDBC-\uDDCC\uDDD0\uDDE0-\uDDFF\uDE70-\uDE7C\uDE80-\uDE89\uDE8F-\uDEC2\uDEC6\uDECE-\uDEDC\uDEDF-\uDEE9]|\uDD3C(?:\u200D[\u2640\u2642]\uFE0F?|\uD83C[\uDFFB-\uDFFF])?|\uDDCE(?:\uD83C[\uDFFB-\uDFFF])?(?:\u200D(?:[\u2640\u2642]\uFE0F?(?:\u200D\u27A1\uFE0F?)?|\u27A1\uFE0F?))?|\uDDD1(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\uD83C[\uDF3E\uDF73\uDF7C\uDF84\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83E\uDDD1|\uDDD1\u200D\uD83E\uDDD2(?:\u200D\uD83E\uDDD2)?|\uDDD2(?:\u200D\uD83E\uDDD2)?))|\uD83C(?:\uDFFB(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D(?:\uD83D\uDC8B\u200D)?\uD83E\uDDD1\uD83C[\uDFFC-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF84\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83E\uDDD1\uD83C[\uDFFB-\uDFFF])))?|\uDFFC(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D(?:\uD83D\uDC8B\u200D)?\uD83E\uDDD1\uD83C[\uDFFB\uDFFD-\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF84\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83E\uDDD1\uD83C[\uDFFB-\uDFFF])))?|\uDFFD(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D(?:\uD83D\uDC8B\u200D)?\uD83E\uDDD1\uD83C[\uDFFB\uDFFC\uDFFE\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF84\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83E\uDDD1\uD83C[\uDFFB-\uDFFF])))?|\uDFFE(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D(?:\uD83D\uDC8B\u200D)?\uD83E\uDDD1\uD83C[\uDFFB-\uDFFD\uDFFF]|\uD83C[\uDF3E\uDF73\uDF7C\uDF84\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83E\uDDD1\uD83C[\uDFFB-\uDFFF])))?|\uDFFF(?:\u200D(?:[\u2695\u2696\u2708]\uFE0F?|\u2764\uFE0F?\u200D(?:\uD83D\uDC8B\u200D)?\uD83E\uDDD1\uD83C[\uDFFB-\uDFFE]|\uD83C[\uDF3E\uDF73\uDF7C\uDF84\uDF93\uDFA4\uDFA8\uDFEB\uDFED]|\uD83D[\uDCBB\uDCBC\uDD27\uDD2C\uDE80\uDE92]|\uD83E(?:[\uDDAF\uDDBC\uDDBD](?:\u200D\u27A1\uFE0F?)?|[\uDDB0-\uDDB3]|\uDD1D\u200D\uD83E\uDDD1\uD83C[\uDFFB-\uDFFF])))?))?|\uDEF1(?:\uD83C(?:\uDFFB(?:\u200D\uD83E\uDEF2\uD83C[\uDFFC-\uDFFF])?|\uDFFC(?:\u200D\uD83E\uDEF2\uD83C[\uDFFB\uDFFD-\uDFFF])?|\uDFFD(?:\u200D\uD83E\uDEF2\uD83C[\uDFFB\uDFFC\uDFFE\uDFFF])?|\uDFFE(?:\u200D\uD83E\uDEF2\uD83C[\uDFFB-\uDFFD\uDFFF])?|\uDFFF(?:\u200D\uD83E\uDEF2\uD83C[\uDFFB-\uDFFE])?))?)/g.test(character) ? width += 2 : width += eastAsianWidth(codePoint, eastAsianWidthOptions)));
    }
    return width;
}

const stdinDiscarder = new class {
    #activeCount=0;
    start() {
        this.#activeCount++, 1 === this.#activeCount && this.#realStart();
    }
    stop() {
        if (this.#activeCount <= 0) throw new Error("`stop` called more times than `start`");
        this.#activeCount--, 0 === this.#activeCount && this.#realStop();
    }
    #realStart() {
        "win32" !== process$2.platform && process$2.stdin.isTTY && (process$2.stdin.setRawMode(!0), 
        process$2.stdin.on("data", this.#handleInput), process$2.stdin.resume());
    }
    #realStop() {
        process$2.stdin.isTTY && (process$2.stdin.off("data", this.#handleInput), process$2.stdin.pause(), 
        process$2.stdin.setRawMode(!1));
    }
    #handleInput(chunk) {
        3 === chunk[0] && process$2.emit("SIGINT");
    }
};

class Ora {
    #linesToClear=0;
    #isDiscardingStdin=!1;
    #lineCount=0;
    #frameIndex=0;
    #options;
    #spinner;
    #stream;
    #id;
    #initialInterval;
    #isEnabled;
    #isSilent;
    #indent;
    #text;
    #prefixText;
    #suffixText;
    color;
    constructor(options) {
        "string" == typeof options && (options = {
            text: options
        }), this.#options = {
            color: "cyan",
            stream: process$2.stderr,
            discardStdin: !0,
            hideCursor: !0,
            ...options
        }, this.color = this.#options.color, this.spinner = this.#options.spinner, this.#initialInterval = this.#options.interval, 
        this.#stream = this.#options.stream, this.#isEnabled = "boolean" == typeof this.#options.isEnabled ? this.#options.isEnabled : function({stream: stream = process.stdout} = {}) {
            return Boolean(stream && stream.isTTY && "dumb" !== process.env.TERM && !("CI" in process.env));
        }({
            stream: this.#stream
        }), this.#isSilent = "boolean" == typeof this.#options.isSilent && this.#options.isSilent, 
        this.text = this.#options.text, this.prefixText = this.#options.prefixText, this.suffixText = this.#options.suffixText, 
        this.indent = this.#options.indent, "test" === process$2.env.NODE_ENV && (this._stream = this.#stream, 
        this._isEnabled = this.#isEnabled, Object.defineProperty(this, "_linesToClear", {
            get() {
                return this.#linesToClear;
            },
            set(newValue) {
                this.#linesToClear = newValue;
            }
        }), Object.defineProperty(this, "_frameIndex", {
            get() {
                return this.#frameIndex;
            }
        }), Object.defineProperty(this, "_lineCount", {
            get() {
                return this.#lineCount;
            }
        }));
    }
    get indent() {
        return this.#indent;
    }
    set indent(indent = 0) {
        if (!(indent >= 0 && Number.isInteger(indent))) throw new Error("The `indent` option must be an integer from 0 and up");
        this.#indent = indent, this.#updateLineCount();
    }
    get interval() {
        return this.#initialInterval ?? this.#spinner.interval ?? 100;
    }
    get spinner() {
        return this.#spinner;
    }
    set spinner(spinner) {
        if (this.#frameIndex = 0, this.#initialInterval = void 0, "object" == typeof spinner) {
            if (void 0 === spinner.frames) throw new Error("The given spinner must have a `frames` property");
            this.#spinner = spinner;
        } else if (function() {
            const {env: env} = process$2, {TERM: TERM, TERM_PROGRAM: TERM_PROGRAM} = env;
            return "win32" !== process$2.platform ? "linux" !== TERM : Boolean(env.WT_SESSION) || Boolean(env.TERMINUS_SUBLIME) || "{cmd::Cmder}" === env.ConEmuTask || "Terminus-Sublime" === TERM_PROGRAM || "vscode" === TERM_PROGRAM || "xterm-256color" === TERM || "alacritty" === TERM || "rxvt-unicode" === TERM || "rxvt-unicode-256color" === TERM || "JetBrains-JediTerm" === env.TERMINAL_EMULATOR;
        }()) if (void 0 === spinner) this.#spinner = cliSpinners.dots; else {
            if ("default" === spinner || !cliSpinners[spinner]) throw new Error(`There is no built-in spinner named '${spinner}'. See https://github.com/sindresorhus/cli-spinners/blob/main/spinners.json for a full list.`);
            this.#spinner = cliSpinners[spinner];
        } else this.#spinner = cliSpinners.line;
    }
    get text() {
        return this.#text;
    }
    set text(value = "") {
        this.#text = value, this.#updateLineCount();
    }
    get prefixText() {
        return this.#prefixText;
    }
    set prefixText(value = "") {
        this.#prefixText = value, this.#updateLineCount();
    }
    get suffixText() {
        return this.#suffixText;
    }
    set suffixText(value = "") {
        this.#suffixText = value, this.#updateLineCount();
    }
    get isSpinning() {
        return void 0 !== this.#id;
    }
    #getFullPrefixText(prefixText = this.#prefixText, postfix = " ") {
        return "string" == typeof prefixText && "" !== prefixText ? prefixText + postfix : "function" == typeof prefixText ? prefixText() + postfix : "";
    }
    #getFullSuffixText(suffixText = this.#suffixText, prefix = " ") {
        return "string" == typeof suffixText && "" !== suffixText ? prefix + suffixText : "function" == typeof suffixText ? prefix + suffixText() : "";
    }
    #updateLineCount() {
        const columns = this.#stream.columns ?? 80, fullPrefixText = this.#getFullPrefixText(this.#prefixText, "-"), fullSuffixText = this.#getFullSuffixText(this.#suffixText, "-"), fullText = " ".repeat(this.#indent) + fullPrefixText + "--" + this.#text + "--" + fullSuffixText;
        this.#lineCount = 0;
        for (const line of stripAnsi(fullText).split("\n")) this.#lineCount += Math.max(1, Math.ceil(stringWidth(line, {
            countAnsiEscapeCodes: !0
        }) / columns));
    }
    get isEnabled() {
        return this.#isEnabled && !this.#isSilent;
    }
    set isEnabled(value) {
        if ("boolean" != typeof value) throw new TypeError("The `isEnabled` option must be a boolean");
        this.#isEnabled = value;
    }
    get isSilent() {
        return this.#isSilent;
    }
    set isSilent(value) {
        if ("boolean" != typeof value) throw new TypeError("The `isSilent` option must be a boolean");
        this.#isSilent = value;
    }
    frame() {
        const {frames: frames} = this.#spinner;
        let frame = frames[this.#frameIndex];
        this.color && (frame = chalk[this.color](frame)), this.#frameIndex = ++this.#frameIndex % frames.length;
        return ("string" == typeof this.#prefixText && "" !== this.#prefixText ? this.#prefixText + " " : "") + frame + ("string" == typeof this.text ? " " + this.text : "") + ("string" == typeof this.#suffixText && "" !== this.#suffixText ? " " + this.#suffixText : "");
    }
    clear() {
        if (!this.#isEnabled || !this.#stream.isTTY) return this;
        this.#stream.cursorTo(0);
        for (let index = 0; index < this.#linesToClear; index++) index > 0 && this.#stream.moveCursor(0, -1), 
        this.#stream.clearLine(1);
        return (this.#indent || this.lastIndent !== this.#indent) && this.#stream.cursorTo(this.#indent), 
        this.lastIndent = this.#indent, this.#linesToClear = 0, this;
    }
    render() {
        return this.#isSilent || (this.clear(), this.#stream.write(this.frame()), this.#linesToClear = this.#lineCount), 
        this;
    }
    start(text) {
        return text && (this.text = text), this.#isSilent ? this : this.#isEnabled ? (this.isSpinning || (this.#options.hideCursor && cliCursor.hide(this.#stream), 
        this.#options.discardStdin && process$2.stdin.isTTY && (this.#isDiscardingStdin = !0, 
        stdinDiscarder.start()), this.render(), this.#id = setInterval(this.render.bind(this), this.interval)), 
        this) : (this.text && this.#stream.write(`- ${this.text}\n`), this);
    }
    stop() {
        return this.#isEnabled ? (clearInterval(this.#id), this.#id = void 0, this.#frameIndex = 0, 
        this.clear(), this.#options.hideCursor && cliCursor.show(this.#stream), this.#options.discardStdin && process$2.stdin.isTTY && this.#isDiscardingStdin && (stdinDiscarder.stop(), 
        this.#isDiscardingStdin = !1), this) : this;
    }
    succeed(text) {
        return this.stopAndPersist({
            symbol: logSymbols.success,
            text: text
        });
    }
    fail(text) {
        return this.stopAndPersist({
            symbol: logSymbols.error,
            text: text
        });
    }
    warn(text) {
        return this.stopAndPersist({
            symbol: logSymbols.warning,
            text: text
        });
    }
    info(text) {
        return this.stopAndPersist({
            symbol: logSymbols.info,
            text: text
        });
    }
    stopAndPersist(options = {}) {
        if (this.#isSilent) return this;
        const prefixText = options.prefixText ?? this.#prefixText, fullPrefixText = this.#getFullPrefixText(prefixText, " "), symbolText = options.symbol ?? " ", text = options.text ?? this.text, fullText = "string" == typeof text ? " " + text : "", suffixText = options.suffixText ?? this.#suffixText, textToWrite = fullPrefixText + symbolText + fullText + this.#getFullSuffixText(suffixText, " ") + "\n";
        return this.stop(), this.#stream.write(textToWrite), this;
    }
}

function ora(options) {
    return new Ora(options);
}

async function installRust({version: version} = {}) {
    const spinner = ora("Installing the rust toolchain using rustup").start();
    try {
        appendPathToRCFiles(TOOL_CONFIG.rust.pathSource, "rust");
        if ("linux" == detectOperatingSystem() && await checkCommand("apt --version")) {
            spinner.text = "Checking for required linux dependencies";
            const missingDeps = await async function(exitWhenMissing = !0, printInstallCommand = !0) {
                const deps = [ "build-essential", "pkg-config", "libudev-dev", "llvm", "libclang-dev", "protobuf-compiler", "libssl-dev" ], statuses = await Promise.allSettled(deps.map((dep => checkCommand(`apt list ${dep} | grep [installed]`, {
                    exit: !1
                })))), missingDeps = [];
                if (statuses.map(((res, index) => {
                    let isInstalled = !1;
                    "rejected" != res.status && res?.value ? res.value.split("\n").map((data => {
                        new RegExp(`^${deps[index]}/[^\\s]+.+\\[installed\\]`, "m").test(data) && (isInstalled = !0);
                    })) : console.error("Unable to detect dependency:", deps[index]), isInstalled || missingDeps.push(deps[index]);
                })), 0 != missingDeps.length) return printInstallCommand && console.log("sudo apt update && sudo apt install -y", missingDeps.join(" ")), 
                exitWhenMissing && (console.error("Missing dependencies:", missingDeps.join(" ")), 
                process.exit(0)), missingDeps;
            }(!0, !0);
            if (missingDeps && missingDeps.length > 0) throw "Your system is missing required system dependencies: " + missingDeps.join(" ");
        }
        let installedVersion = await installedToolVersion("rust");
        if (installedVersion) return spinner.info(`rust ${installedVersion} is already installed`), 
        !0;
        spinner.text = "Installing the rust toolchain using rustup";
        const result = await src_default("curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y");
        if (result && 0 != result.code) {
            throw result.stderr.trim().split("\n").join("\n");
        }
        return spinner.text = "Verifying rust was installed", installedVersion = await installedToolVersion("rust"), 
        installedVersion ? (spinner.succeed(`rust ${installedVersion} installed`), installedVersion) : (spinner.fail("rust failed to install"), 
        !1);
    } catch (err) {
        spinner.fail("Unable to install rust"), "string" == typeof err ? console.error(err) : (Error, 
        console.error(err.message));
    }
    return !1;
}

async function isDockerInstalled() {
    return checkCommand("docker --version", {
        exit: !1,
        onError: () => {
            console.error("Unable to detect Docker (which is required for 'solana-verify'). Do you have it installed?"), 
            console.error("To install Docker, follow the instructions in the official Docker documentation:", "\nhttps://docs.docker.com/engine/install/");
        }
    });
}

async function checkShellPathSource(cmd, pathSource) {
    const [withoutPath, withPath] = await Promise.allSettled([ src_default(cmd), src_default(`export PATH="${pathSource}:$PATH" && ${cmd}`) ]);
    if ("fulfilled" == withoutPath.status && "fulfilled" == withPath.status) {
        if (0 === withoutPath.value.code) return withoutPath.value.stdout.toLowerCase() == withPath.value.stdout.toLowerCase() ? PathSourceStatus.SUCCESS : PathSourceStatus.OUTPUT_MISMATCH;
        if (0 === withPath.value.code) return PathSourceStatus.MISSING_PATH;
    }
    return PathSourceStatus.FAILED;
}

const toolNames = [ "rust", "solana", "avm", "mucho", "anchor", "trident", "zest", "yarn", "verify" ];

function installCommand() {
    return new Command("install").configureOutput(cliOutputConfig).description("install Solana development tooling").addArgument(new Argument("<tool>", "tool to install (default: all)").choices(toolNames).argOptional()).addArgument(new Argument("<version>", "desired tool version to install (default: stable)").argOptional()).action((async (toolName, version, options) => {
        var msg;
        msg = "Install Solana development tooling", console.log(picocolors.inverse(` ${msg} `));
        const os = detectOperatingSystem();
        "windows" == os && warnMessage("Windows is not yet natively supported for the rust based tooling.\nWe recommend using WSL inside your Windows terminal."), 
        await async function({outputToolStatus: outputToolStatus = !1} = {}) {
            let allInstalled = !0, status = {
                rust: !1,
                solana: !1,
                avm: !1,
                anchor: !1,
                yarn: !1,
                mucho: !1,
                trident: !1,
                zest: !1,
                verify: !1
            };
            if (await Promise.all(Object.keys(status).map((async tool => {
                const version = await installedToolVersion(tool);
                status[tool] = version, status[tool] || (allInstalled = !1);
            }))), outputToolStatus) {
                let noteContent = "";
                for (const command in status) Object.prototype.hasOwnProperty.call(status, command) && (noteContent += "- ", 
                status[command] ? noteContent += picocolors.green(command) : noteContent += picocolors.red(command), 
                noteContent += ` ${status[command] || "(not installed)"}\n`);
                console.log(noteContent.trim(), "\n");
            }
            return {
                allInstalled: allInstalled,
                status: status
            };
        }({
            outputToolStatus: !1
        });
        const pathsToRefresh = [];
        toolName && "rust" != toolName || (await installRust({
            os: os,
            version: version
        }), await checkShellPathSource(TOOL_CONFIG.rust.version, TOOL_CONFIG.rust.pathSource).then((status => status != PathSourceStatus.MISSING_PATH || pathsToRefresh.push(TOOL_CONFIG.rust.pathSource)))), 
        toolName && "solana" != toolName || (await async function({version: version = "stable"} = {}) {
            const spinner = ora("Installing the Solana CLI tool suite...").start();
            try {
                appendPathToRCFiles(TOOL_CONFIG.solana.pathSource, "solana");
                let installedVersion = await installedToolVersion("solana");
                if (installedVersion) return spinner.info(`solana ${installedVersion} is already installed`), 
                !0;
                if ("stable" != (version = version.toLowerCase()) && !version.startsWith("v")) throw `Invalid version: '${version}'`;
                spinner.text = "Installing the Solana CLI tool suite...";
                const result = await src_default(`sh -c "$(curl -sSfL https://release.anza.xyz/${version}/install)"`);
                if (result && 0 != result.code) throw result.stderr.trim().split("\n").join("\n");
                return spinner.text = "Verifying solana was installed", installedVersion = await installedToolVersion("solana"), 
                installedVersion ? (spinner.succeed(`solana ${installedVersion} installed`), installedVersion) : (spinner.fail("solana failed to install"), 
                !1);
            } catch (err) {
                spinner.fail("Unable to install the Solana CLI tool suite"), "string" == typeof err ? console.error(err) : (Error, 
                console.error(err.message));
            }
            return !1;
        }({
            os: os,
            version: version
        }), await checkShellPathSource(TOOL_CONFIG.solana.version, TOOL_CONFIG.solana.pathSource).then((status => status != PathSourceStatus.MISSING_PATH || pathsToRefresh.push(TOOL_CONFIG.solana.pathSource)))), 
        toolName && "mucho" != toolName || await async function({} = {}) {
            const spinner = ora("Installing the mucho cli...").start();
            try {
                let installedVersion = await installedToolVersion("mucho");
                return installedVersion ? (spinner.info(`mucho ${installedVersion} is already installed`), 
                !0) : (spinner.text = "Installing the2 mucho cli", await src_default("npm install -g mucho"), 
                spinner.text = "Verifying mucho was installed", installedVersion = await installedToolVersion("mucho"), 
                installedVersion ? (spinner.succeed(`mucho ${installedVersion} installed`), installedVersion) : (spinner.fail("mucho cli failed to install"), 
                !1));
            } catch (err) {
                spinner.fail("Unable to install the mucho cli"), "string" == typeof err ? console.error(err) : (Error, 
                console.error(err.message));
            }
            return !1;
        }({
            os: os,
            version: version
        }), toolName && "avm" != toolName || await async function({version: version = "latest"} = {}) {
            const spinner = ora("Installing avm (anchor version manager)").start();
            try {
                let installedVersion = await installedToolVersion("avm");
                if (installedVersion) return spinner.info(`avm ${installedVersion} is already installed`), 
                !0;
                spinner.text = "Installing avm (anchor version manager)";
                const result = await src_default("cargo install --git https://github.com/coral-xyz/anchor avm --locked --force");
                if (result && 0 != result.code) throw result.stderr.trim().split("\n").join("\n");
                return spinner.text = "Verifying avm was installed", installedVersion = await installedToolVersion("avm"), 
                installedVersion ? (spinner.succeed(`avm ${installedVersion} installed`), installedVersion) : (spinner.fail("avm failed to install"), 
                !1);
            } catch (err) {
                spinner.fail("Unable to install avm"), "string" == typeof err ? console.error(err) : (Error, 
                console.error(err.message));
            }
            return !1;
        }({
            os: os,
            version: version
        }), toolName && "anchor" != toolName || await async function({verifyParentCommand: verifyParentCommand = !0, version: version = "latest"} = {}) {
            const spinner = ora("Installing anchor using avm...").start();
            try {
                let installedVersion = await installedToolVersion("anchor");
                if (installedVersion && installedVersion == version) return spinner.info(`anchor ${installedVersion} is already installed`), 
                !0;
                spinner.text = "Verifying avm is installed";
                const avmVersion = await installedToolVersion("avm");
                if (!avmVersion) throw "avm is NOT already installed";
                if (version = avmVersion, installedVersion && installedVersion == version) return spinner.info(`anchor ${installedVersion} is already installed`), 
                !0;
                let result;
                try {
                    spinner.text = `Installing anchor version '${version}'. This may take a few moments...`, 
                    result = await src_default(`avm install ${version}`);
                } catch (err) {
                    throw "Unable to execute `avm install`";
                }
                if (result && 0 != result.code) throw result.stderr.trim().split("\n").join("\n");
                try {
                    spinner.text = "Setting anchor version with avm", result = await src_default(`avm use ${version}`);
                } catch (err) {
                    throw "Unable to execute `avm use`";
                }
                return spinner.text = "Verifying anchor was installed", installedVersion = await installedToolVersion("anchor"), 
                installedVersion ? (spinner.succeed(`anchor ${installedVersion} installed using avm`), 
                installedVersion) : (spinner.fail("anchor failed to install"), !1);
            } catch (err) {
                spinner.fail("Unable to install anchor using avm"), "string" == typeof err ? console.error(err) : (Error, 
                console.error(err.message));
            }
            return !1;
        }({
            os: os,
            version: version
        }), toolName && "trident" != toolName || await async function({version: version = "latest", verifyParentCommand: verifyParentCommand = !0} = {}) {
            const spinner = ora("Installing trident (fuzzer)").start();
            try {
                let installedVersion = await installedToolVersion("trident");
                if (installedVersion) return spinner.info(`trident ${installedVersion} is already installed`), 
                !0;
                if (verifyParentCommand && !await installedToolVersion("rust")) throw spinner.fail("Rust/cargo was not found"), 
                "parent command not found";
                spinner.text = "Installing trident (fuzzer)";
                const result = await src_default("cargo install honggfuzz trident-cli");
                if (result && 0 != result.code) throw result.stderr.trim().split("\n").join("\n");
                return spinner.text = "Verifying trident was installed", installedVersion = await installedToolVersion("trident"), 
                installedVersion ? (spinner.succeed(`trident ${installedVersion} installed`), installedVersion) : (spinner.fail("trident failed to install"), 
                !1);
            } catch (err) {
                spinner.fail("Unable to install the trident fuzzer"), "string" == typeof err ? console.error(err) : (Error, 
                console.error(err.message));
            }
            return !1;
        }({
            os: os,
            version: version
        }), toolName && "zest" != toolName || await async function({version: version = "latest", verifyParentCommand: verifyParentCommand = !0} = {}) {
            const spinner = ora("Installing zest (code coverage)").start();
            try {
                let result, installedVersion = await installedToolVersion("zest");
                if (installedVersion) return spinner.info(`zest ${installedVersion} is already installed`), 
                !0;
                if (verifyParentCommand && !await installedToolVersion("rust")) throw "rustc/cargo was not found but is required";
                try {
                    spinner.text = "Installing zest (code coverage)...", result = await src_default("cargo install --git https://github.com/LimeChain/zest zest --force");
                } catch (err) {
                    throw "Unable to execute the zest installer";
                }
                if (result && 0 != result.code) {
                    const error = result.stderr.trim().split("\n");
                    let parsed = null;
                    if (parsed = error.slice(-1)[0].match(/(?<=\bit requires rustc\s)\d+\.\d+(\.\d+)?(?=\sor newer\b)/gi)[0]) throw `Zest requires a minimum rustc version of ${parsed}. To set your rustc version, run the following command:\nrustup default ${parsed}`;
                    throw error.join("\n");
                }
                try {
                    spinner.text = "Installing the 'llvm-tools-preview' component in rustup", result = await src_default("rustup component add llvm-tools-preview");
                } catch (err) {
                    throw "Unable to execute 'rustup component add llvm-tools-preview'. Try running manually.";
                }
                return spinner.text = "Verifying zest was installed", installedVersion = await installedToolVersion("zest"), 
                installedVersion ? (spinner.succeed(`zest ${installedVersion} installed`), installedVersion) : (spinner.fail("zest failed to install"), 
                !1);
            } catch (err) {
                spinner.fail("Unable to install zest"), "string" == typeof err ? console.error(err) : (Error, 
                console.error(err.message));
            }
            return !1;
        }({
            os: os,
            version: version
        }), toolName && "verify" != toolName || await async function({version: version = "latest", verifyParentCommand: verifyParentCommand = !0} = {}) {
            const spinner = ora("Installing solana-verify").start();
            try {
                let result, installedVersion = await installedToolVersion("verify");
                if (installedVersion) return spinner.info(`solana-verify ${installedVersion} is already installed`), 
                await isDockerInstalled(), !0;
                if (verifyParentCommand && !await installedToolVersion("rust")) throw "rustc/cargo was not found but is required";
                try {
                    spinner.text = "Installing solana-verify...", result = await src_default("cargo install solana-verify");
                } catch (err) {
                    throw "Unable to execute the solana-verify installer";
                }
                if (result && 0 != result.code) {
                    throw result.stderr.trim().split("\n").join("\n");
                }
                return spinner.text = "Verifying solana-verify was installed", installedVersion = await installedToolVersion("verify"), 
                installedVersion ? (spinner.succeed(`solana-verify ${installedVersion} installed`), 
                await isDockerInstalled(), installedVersion) : (spinner.fail("solana-verify failed to install"), 
                !1);
            } catch (err) {
                spinner.fail("Unable to install solana-verify"), "string" == typeof err ? console.error(err) : (Error, 
                console.error(err.message));
            }
            return !1;
        }({
            os: os,
            version: version
        }), toolName && "yarn" != toolName || await async function({} = {}) {
            const spinner = ora("Installing yarn package manager...").start();
            try {
                let installedVersion = await installedToolVersion("yarn");
                return installedVersion ? (spinner.info(`yarn ${installedVersion} is already installed`), 
                !0) : (spinner.text = "Installing yarn package manager", await src_default("npm install -g yarn"), 
                spinner.text = "Verifying yarn was installed", installedVersion = await installedToolVersion("yarn"), 
                installedVersion ? (spinner.succeed(`yarn ${installedVersion} installed`), installedVersion) : (spinner.fail("yarn package manager failed to install"), 
                !1));
            } catch (err) {
                spinner.fail("Unable to install yarn package manager"), "string" == typeof err ? console.error(err) : (Error, 
                console.error(err.message));
            }
            return !1;
        }({
            os: os,
            version: version
        }), pathsToRefresh.length > 0 && (console.log("\nClose and reopen your terminal to apply the required", "PATH changes \nor run the following in your existing shell:", "\n"), 
        console.log(`export PATH="${pathsToRefresh.join(":")}:$PATH"`, "\n"));
    }));
}

"undefined" != typeof Bun || function(currentVersion, requiredVersion) {
    const [major, minor] = currentVersion.split(".").map(Number), [reqMajor, reqMinor] = requiredVersion.split(".").map(Number);
    return major > reqMajor || major === reqMajor && minor >= reqMinor;
}(process.versions.node, "22.0.0") || (console.error("This tool requires Node.js v22.0.0 or higher.", `You are using v${process.versions.node}.`), 
process.exit(1)), async function() {
    try {
        const program = cliProgramRoot();
        program.addCommand(installCommand()), 2 === process.argv.length && process.argv.push("--help"), 
        await program.parseAsync();
    } catch (err) {
        errorMessage(err.toString());
    }
}();
